// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/apis/BasketsApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasketsApiResponseProcessor = exports.BasketsApiRequestFactory = void 0;
// TODO: better import syntax?
const baseapi_1 = require("./baseapi");
const http_1 = require("../http/http");
const ObjectSerializer_1 = require("../models/ObjectSerializer");
const exception_1 = require("./exception");
const util_1 = require("../util");
/**
 * no description
 */
class BasketsApiRequestFactory extends baseapi_1.BaseAPIRequestFactory {
    /**
     * This adds a package (an object describing the product) to the basket `{ident}`. For subscriptions, **only one subscription item may be in a basket at a time**, and it cannot be included with one-time payment items. **This endpoint requires prior approval. Please contact your account manager.**
     * Add a package to the basket
     * @param ident The basket identifier.
     * @param addPackageRequest
     */
    async addPackage(ident, addPackageRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'ident' is not null or undefined
        if (ident === null || ident === undefined) {
            throw new baseapi_1.RequiredError("BasketsApi", "addPackage", "ident");
        }
        // Path Params
        const localVarPath = '/baskets/{ident}/packages'
            .replace('{' + 'ident' + '}', encodeURIComponent(String(ident)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(addPackageRequest, "AddPackageRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Adds a `Sale` to the basket with `{ident}`. **Sales cannot be applied to baskets with `revenue_share` set.**
     * Add a sale to the basket
     * @param ident The basket identifier.
     * @param sale Provide a &#x60;Sale&#x60; as an object to apply it to the basket.
     */
    async addSaleToBasket(ident, sale, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'ident' is not null or undefined
        if (ident === null || ident === undefined) {
            throw new baseapi_1.RequiredError("BasketsApi", "addSaleToBasket", "ident");
        }
        // Path Params
        const localVarPath = '/baskets/{ident}/sales'
            .replace('{' + 'ident' + '}', encodeURIComponent(String(ident)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(sale, "Sale", ""), contentType);
        requestContext.setBody(serializedBody);
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * This will create and return a `Basket` that can be paid for by redirecting the user to `links.checkout`
     * Create a basket that can be used to pay for items
     * @param createBasketRequest Create a basket, returning the full basket object and payment link.
     */
    async createBasket(createBasketRequest, _options) {
        let _config = _options || this.configuration;
        // Path Params
        const localVarPath = '/baskets';
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(createBasketRequest, "CreateBasketRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets the basket associated with the provided identifier.
     * Fetch a basket by its identifier
     * @param ident The basket identifier.
     */
    async getBasketById(ident, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'ident' is not null or undefined
        if (ident === null || ident === undefined) {
            throw new baseapi_1.RequiredError("BasketsApi", "getBasketById", "ident");
        }
        // Path Params
        const localVarPath = '/baskets/{ident}'
            .replace('{' + 'ident' + '}', encodeURIComponent(String(ident)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * This will remove the given `{rows.id}` from the basket `{ident}`. The basket must be re-fetched after running to receive updated totals.
     * Remove a row from the basket
     * @param ident The basket identifier.
     * @param rowsId The &#x60;id&#x60; of the &#x60;basket.rows&#x60; row to remove.
     */
    async removeRowFromBasket(ident, rowsId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'ident' is not null or undefined
        if (ident === null || ident === undefined) {
            throw new baseapi_1.RequiredError("BasketsApi", "removeRowFromBasket", "ident");
        }
        // verify required parameter 'rowsId' is not null or undefined
        if (rowsId === null || rowsId === undefined) {
            throw new baseapi_1.RequiredError("BasketsApi", "removeRowFromBasket", "rowsId");
        }
        // Path Params
        const localVarPath = '/baskets/{ident}/packages/{rows.id}'
            .replace('{' + 'ident' + '}', encodeURIComponent(String(ident)))
            .replace('{' + 'rows.id' + '}', encodeURIComponent(String(rowsId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.DELETE);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
}
exports.BasketsApiRequestFactory = BasketsApiRequestFactory;
class BasketsApiResponseProcessor {
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to addPackage
     * @throws ApiException if the response code was not in [200, 299]
     */
    async addPackageWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("400", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Improperly formatted package. See ErrorResponse.", undefined, response.headers);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Basket not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to addSaleToBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async addSaleToBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("400", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Bad request. Improperly formatted Sale or this basket cannot accept sales. See ErrorResponse.", undefined, response.headers);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Basket not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to createBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async createBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getBasketById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getBasketByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Basket not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to removeRowFromBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async removeRowFromBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("204", response.httpStatusCode)) {
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, undefined);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Row or basket not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "void", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
}
exports.BasketsApiResponseProcessor = BasketsApiResponseProcessor;
