// @tebexio/tebex-sdk-nodejs/dist/Checkout/checkout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Checkout = void 0;
const lib_1 = require("./lib");
const basketBuilder_1 = require("./basketBuilder");
const packageBuilder_1 = require("./packageBuilder");
class Checkout {
    constructor(projectId, privateKey) {
        let auth = (0, lib_1.createConfiguration)({
            authMethods: {
                tebex_checkout_auth_basic: {
                    username: projectId,
                    password: privateKey
                }
            },
        });
        this.checkoutApi = new lib_1.CheckoutApi(auth);
        this.paymentsApi = new lib_1.PaymentsApi(auth);
        this.recurringPaymentsApi = new lib_1.RecurringPaymentsApi(auth);
        this.basketsApi = new lib_1.BasketsApi(auth);
    }
    setApiKeys(projectId, privateKey) {
        return new Checkout(projectId, privateKey);
    }
    newBasketBuilder() {
        return new basketBuilder_1.BasketBuilder();
    }
    newPackageBuilder() {
        return new packageBuilder_1.PackageBuilder();
    }
    createBasket(createRequest) {
        return this.basketsApi.createBasket(createRequest);
    }
    addPackage(basket, pack) {
        return this.basketsApi.addPackage(basket.ident, {
            _package: pack,
            qty: pack.qty,
            type: pack.type == lib_1.PackageTypeEnum.Single ? lib_1.AddPackageRequestTypeEnum.Single : lib_1.AddPackageRequestTypeEnum.Subscription
        });
    }
    checkoutRequest(basketBuilder, checkoutItems, sale) {
        return this.checkoutApi.checkout({
            basket: basketBuilder.buildCheckoutRequest(),
            items: checkoutItems,
            sale: sale
        });
    }
    getBasket(ident) {
        return this.basketsApi.getBasketById(ident);
    }
    removeBasketRow(basket, rowId) {
        return this.basketsApi.removeRowFromBasket(basket.ident, rowId);
    }
    addSaleToBasket(basket, saleName, saleType, saleAmount) {
        return this.basketsApi.addSaleToBasket(basket.ident, {
            name: saleName,
            discountType: saleType,
            amount: saleAmount,
        });
    }
    getPayment(identifier) {
        return this.paymentsApi.getPaymentById(identifier);
    }
    refundPayment(transactionId) {
        return this.paymentsApi.refundPaymentById(transactionId);
    }
    getRecurringPayment(transactionId) {
        return this.recurringPaymentsApi.getRecurringPayment(transactionId);
    }
    updateSubscriptionProduct(recurringPaymentReference, packages) {
        let items = [];
        packages.forEach(p => {
            items.push({
                type: p.type == lib_1.PackageTypeEnum.Single ? lib_1.UpdateSubscriptionRequestItemsInnerTypeEnum.Single : lib_1.UpdateSubscriptionRequestItemsInnerTypeEnum.Subscription,
                qty: p.qty,
                _package: p,
            });
        });
        return this.recurringPaymentsApi.updateSubscription(recurringPaymentReference, {
            items: items
        });
    }
    cancelRecurringPayment(recurringPaymentReference) {
        return this.recurringPaymentsApi.cancelRecurringPayment(recurringPaymentReference);
    }
    pauseRecurringPayment(recurringPaymentReference) {
        return this.recurringPaymentsApi.updateRecurringPayment(recurringPaymentReference, {
            status: lib_1.UpdateRecurringPaymentRequestStatusEnum.Paused
        });
    }
    reactivateRecurringPayment(recurringPaymentReference) {
        return this.recurringPaymentsApi.updateRecurringPayment(recurringPaymentReference, {
            status: lib_1.UpdateRecurringPaymentRequestStatusEnum.Active
        });
    }
}
exports.Checkout = Checkout;
