// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/types/ObjectParamAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectRecurringPaymentsApi = exports.ObjectPaymentsApi = exports.ObjectCheckoutApi = exports.ObjectBasketsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class ObjectBasketsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableBasketsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This adds a package (an object describing the product) to the basket `{ident}`. For subscriptions, **only one subscription item may be in a basket at a time**, and it cannot be included with one-time payment items. **This endpoint requires prior approval. Please contact your account manager.**
     * Add a package to the basket
     * @param param the request object
     */
    addPackageWithHttpInfo(param, options) {
        return this.api.addPackageWithHttpInfo(param.ident, param.addPackageRequest, options).toPromise();
    }
    /**
     * This adds a package (an object describing the product) to the basket `{ident}`. For subscriptions, **only one subscription item may be in a basket at a time**, and it cannot be included with one-time payment items. **This endpoint requires prior approval. Please contact your account manager.**
     * Add a package to the basket
     * @param param the request object
     */
    addPackage(param, options) {
        return this.api.addPackage(param.ident, param.addPackageRequest, options).toPromise();
    }
    /**
     * Adds a `Sale` to the basket with `{ident}`. **Sales cannot be applied to baskets with `revenue_share` set.**
     * Add a sale to the basket
     * @param param the request object
     */
    addSaleToBasketWithHttpInfo(param, options) {
        return this.api.addSaleToBasketWithHttpInfo(param.ident, param.sale, options).toPromise();
    }
    /**
     * Adds a `Sale` to the basket with `{ident}`. **Sales cannot be applied to baskets with `revenue_share` set.**
     * Add a sale to the basket
     * @param param the request object
     */
    addSaleToBasket(param, options) {
        return this.api.addSaleToBasket(param.ident, param.sale, options).toPromise();
    }
    /**
     * This will create and return a `Basket` that can be paid for by redirecting the user to `links.checkout`
     * Create a basket that can be used to pay for items
     * @param param the request object
     */
    createBasketWithHttpInfo(param = {}, options) {
        return this.api.createBasketWithHttpInfo(param.createBasketRequest, options).toPromise();
    }
    /**
     * This will create and return a `Basket` that can be paid for by redirecting the user to `links.checkout`
     * Create a basket that can be used to pay for items
     * @param param the request object
     */
    createBasket(param = {}, options) {
        return this.api.createBasket(param.createBasketRequest, options).toPromise();
    }
    /**
     * Gets the basket associated with the provided identifier.
     * Fetch a basket by its identifier
     * @param param the request object
     */
    getBasketByIdWithHttpInfo(param, options) {
        return this.api.getBasketByIdWithHttpInfo(param.ident, options).toPromise();
    }
    /**
     * Gets the basket associated with the provided identifier.
     * Fetch a basket by its identifier
     * @param param the request object
     */
    getBasketById(param, options) {
        return this.api.getBasketById(param.ident, options).toPromise();
    }
    /**
     * This will remove the given `{rows.id}` from the basket `{ident}`. The basket must be re-fetched after running to receive updated totals.
     * Remove a row from the basket
     * @param param the request object
     */
    removeRowFromBasketWithHttpInfo(param, options) {
        return this.api.removeRowFromBasketWithHttpInfo(param.ident, param.rowsId, options).toPromise();
    }
    /**
     * This will remove the given `{rows.id}` from the basket `{ident}`. The basket must be re-fetched after running to receive updated totals.
     * Remove a row from the basket
     * @param param the request object
     */
    removeRowFromBasket(param, options) {
        return this.api.removeRowFromBasket(param.ident, param.rowsId, options).toPromise();
    }
}
exports.ObjectBasketsApi = ObjectBasketsApi;
const ObservableAPI_2 = require("./ObservableAPI");
class ObjectCheckoutApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableCheckoutApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This API call allows the complete checkout flow (create basket, add items, add sale) to be made in a single API call, for when the Seller is managing the basket locally. **This endpoint requires prior approval - please contact your account manager.**
     * Create a checkout request
     * @param param the request object
     */
    checkoutWithHttpInfo(param = {}, options) {
        return this.api.checkoutWithHttpInfo(param.checkoutRequest, options).toPromise();
    }
    /**
     * This API call allows the complete checkout flow (create basket, add items, add sale) to be made in a single API call, for when the Seller is managing the basket locally. **This endpoint requires prior approval - please contact your account manager.**
     * Create a checkout request
     * @param param the request object
     */
    checkout(param = {}, options) {
        return this.api.checkout(param.checkoutRequest, options).toPromise();
    }
}
exports.ObjectCheckoutApi = ObjectCheckoutApi;
const ObservableAPI_3 = require("./ObservableAPI");
class ObjectPaymentsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservablePaymentsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This will fetch the given payment associated with this transaction id. Single payment transaction IDs begin with `tbx-`
     * Fetch a payment by its transaction ID
     * @param param the request object
     */
    getPaymentByIdWithHttpInfo(param, options) {
        return this.api.getPaymentByIdWithHttpInfo(param.txnId, options).toPromise();
    }
    /**
     * This will fetch the given payment associated with this transaction id. Single payment transaction IDs begin with `tbx-`
     * Fetch a payment by its transaction ID
     * @param param the request object
     */
    getPaymentById(param, options) {
        return this.api.getPaymentById(param.txnId, options).toPromise();
    }
    /**
     * This will refund the given payment associated with this transaction id.
     * Refund a payment by its transaction ID
     * @param param the request object
     */
    refundPaymentByIdWithHttpInfo(param, options) {
        return this.api.refundPaymentByIdWithHttpInfo(param.txnId, options).toPromise();
    }
    /**
     * This will refund the given payment associated with this transaction id.
     * Refund a payment by its transaction ID
     * @param param the request object
     */
    refundPaymentById(param, options) {
        return this.api.refundPaymentById(param.txnId, options).toPromise();
    }
}
exports.ObjectPaymentsApi = ObjectPaymentsApi;
const ObservableAPI_4 = require("./ObservableAPI");
class ObjectRecurringPaymentsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservableRecurringPaymentsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This cancels the recurring payment for the reference provided. Recurring payment references start with `tbx-r-`
     * Cancel a recurring payment
     * @param param the request object
     */
    cancelRecurringPaymentWithHttpInfo(param, options) {
        return this.api.cancelRecurringPaymentWithHttpInfo(param.reference, options).toPromise();
    }
    /**
     * This cancels the recurring payment for the reference provided. Recurring payment references start with `tbx-r-`
     * Cancel a recurring payment
     * @param param the request object
     */
    cancelRecurringPayment(param, options) {
        return this.api.cancelRecurringPayment(param.reference, options).toPromise();
    }
    /**
     * Fetch a recurring payment (subscription) by its reference
     * @param param the request object
     */
    getRecurringPaymentWithHttpInfo(param, options) {
        return this.api.getRecurringPaymentWithHttpInfo(param.reference, options).toPromise();
    }
    /**
     * Fetch a recurring payment (subscription) by its reference
     * @param param the request object
     */
    getRecurringPayment(param, options) {
        return this.api.getRecurringPayment(param.reference, options).toPromise();
    }
    /**
     * Pause or reactivate a recurring payment
     * @param param the request object
     */
    updateRecurringPaymentWithHttpInfo(param, options) {
        return this.api.updateRecurringPaymentWithHttpInfo(param.reference, param.updateRecurringPaymentRequest, options).toPromise();
    }
    /**
     * Pause or reactivate a recurring payment
     * @param param the request object
     */
    updateRecurringPayment(param, options) {
        return this.api.updateRecurringPayment(param.reference, param.updateRecurringPaymentRequest, options).toPromise();
    }
    /**
     * If the new subscription amount is higher than the existing amount, a pro-rata charge will be made to cover the cost of the new price up until the next billing date.   **This endpoint requires prior approval - please contact your account manager.**
     * Update a subscription with a new product / amount to pay - replacing the existing product
     * @param param the request object
     */
    updateSubscriptionWithHttpInfo(param, options) {
        return this.api.updateSubscriptionWithHttpInfo(param.reference, param.updateSubscriptionRequest, options).toPromise();
    }
    /**
     * If the new subscription amount is higher than the existing amount, a pro-rata charge will be made to cover the cost of the new price up until the next billing date.   **This endpoint requires prior approval - please contact your account manager.**
     * Update a subscription with a new product / amount to pay - replacing the existing product
     * @param param the request object
     */
    updateSubscription(param, options) {
        return this.api.updateSubscription(param.reference, param.updateSubscriptionRequest, options).toPromise();
    }
}
exports.ObjectRecurringPaymentsApi = ObjectRecurringPaymentsApi;
