// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseRecurringPaymentsApi = exports.PromisePaymentsApi = exports.PromiseCheckoutApi = exports.PromiseBasketsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseBasketsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableBasketsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This adds a package (an object describing the product) to the basket `{ident}`. For subscriptions, **only one subscription item may be in a basket at a time**, and it cannot be included with one-time payment items. **This endpoint requires prior approval. Please contact your account manager.**
     * Add a package to the basket
     * @param ident The basket identifier.
     * @param [addPackageRequest]
     */
    addPackageWithHttpInfo(ident, addPackageRequest, _options) {
        const result = this.api.addPackageWithHttpInfo(ident, addPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * This adds a package (an object describing the product) to the basket `{ident}`. For subscriptions, **only one subscription item may be in a basket at a time**, and it cannot be included with one-time payment items. **This endpoint requires prior approval. Please contact your account manager.**
     * Add a package to the basket
     * @param ident The basket identifier.
     * @param [addPackageRequest]
     */
    addPackage(ident, addPackageRequest, _options) {
        const result = this.api.addPackage(ident, addPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * Adds a `Sale` to the basket with `{ident}`. **Sales cannot be applied to baskets with `revenue_share` set.**
     * Add a sale to the basket
     * @param ident The basket identifier.
     * @param [sale] Provide a &#x60;Sale&#x60; as an object to apply it to the basket.
     */
    addSaleToBasketWithHttpInfo(ident, sale, _options) {
        const result = this.api.addSaleToBasketWithHttpInfo(ident, sale, _options);
        return result.toPromise();
    }
    /**
     * Adds a `Sale` to the basket with `{ident}`. **Sales cannot be applied to baskets with `revenue_share` set.**
     * Add a sale to the basket
     * @param ident The basket identifier.
     * @param [sale] Provide a &#x60;Sale&#x60; as an object to apply it to the basket.
     */
    addSaleToBasket(ident, sale, _options) {
        const result = this.api.addSaleToBasket(ident, sale, _options);
        return result.toPromise();
    }
    /**
     * This will create and return a `Basket` that can be paid for by redirecting the user to `links.checkout`
     * Create a basket that can be used to pay for items
     * @param [createBasketRequest] Create a basket, returning the full basket object and payment link.
     */
    createBasketWithHttpInfo(createBasketRequest, _options) {
        const result = this.api.createBasketWithHttpInfo(createBasketRequest, _options);
        return result.toPromise();
    }
    /**
     * This will create and return a `Basket` that can be paid for by redirecting the user to `links.checkout`
     * Create a basket that can be used to pay for items
     * @param [createBasketRequest] Create a basket, returning the full basket object and payment link.
     */
    createBasket(createBasketRequest, _options) {
        const result = this.api.createBasket(createBasketRequest, _options);
        return result.toPromise();
    }
    /**
     * Gets the basket associated with the provided identifier.
     * Fetch a basket by its identifier
     * @param ident The basket identifier.
     */
    getBasketByIdWithHttpInfo(ident, _options) {
        const result = this.api.getBasketByIdWithHttpInfo(ident, _options);
        return result.toPromise();
    }
    /**
     * Gets the basket associated with the provided identifier.
     * Fetch a basket by its identifier
     * @param ident The basket identifier.
     */
    getBasketById(ident, _options) {
        const result = this.api.getBasketById(ident, _options);
        return result.toPromise();
    }
    /**
     * This will remove the given `{rows.id}` from the basket `{ident}`. The basket must be re-fetched after running to receive updated totals.
     * Remove a row from the basket
     * @param ident The basket identifier.
     * @param rowsId The &#x60;id&#x60; of the &#x60;basket.rows&#x60; row to remove.
     */
    removeRowFromBasketWithHttpInfo(ident, rowsId, _options) {
        const result = this.api.removeRowFromBasketWithHttpInfo(ident, rowsId, _options);
        return result.toPromise();
    }
    /**
     * This will remove the given `{rows.id}` from the basket `{ident}`. The basket must be re-fetched after running to receive updated totals.
     * Remove a row from the basket
     * @param ident The basket identifier.
     * @param rowsId The &#x60;id&#x60; of the &#x60;basket.rows&#x60; row to remove.
     */
    removeRowFromBasket(ident, rowsId, _options) {
        const result = this.api.removeRowFromBasket(ident, rowsId, _options);
        return result.toPromise();
    }
}
exports.PromiseBasketsApi = PromiseBasketsApi;
const ObservableAPI_2 = require("./ObservableAPI");
class PromiseCheckoutApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableCheckoutApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This API call allows the complete checkout flow (create basket, add items, add sale) to be made in a single API call, for when the Seller is managing the basket locally. **This endpoint requires prior approval - please contact your account manager.**
     * Create a checkout request
     * @param [checkoutRequest] Provide a &#x60;Basket&#x60;, an array of &#x60;Packages&#x60; to be added to the basket, and an optional &#x60;Sale&#x60; to complete the full checkout flow in one call. **Only one subscription item may be in the basket at a time.**
     */
    checkoutWithHttpInfo(checkoutRequest, _options) {
        const result = this.api.checkoutWithHttpInfo(checkoutRequest, _options);
        return result.toPromise();
    }
    /**
     * This API call allows the complete checkout flow (create basket, add items, add sale) to be made in a single API call, for when the Seller is managing the basket locally. **This endpoint requires prior approval - please contact your account manager.**
     * Create a checkout request
     * @param [checkoutRequest] Provide a &#x60;Basket&#x60;, an array of &#x60;Packages&#x60; to be added to the basket, and an optional &#x60;Sale&#x60; to complete the full checkout flow in one call. **Only one subscription item may be in the basket at a time.**
     */
    checkout(checkoutRequest, _options) {
        const result = this.api.checkout(checkoutRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseCheckoutApi = PromiseCheckoutApi;
const ObservableAPI_3 = require("./ObservableAPI");
class PromisePaymentsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservablePaymentsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This will fetch the given payment associated with this transaction id. Single payment transaction IDs begin with `tbx-`
     * Fetch a payment by its transaction ID
     * @param txnId The payment reference to fetch.
     */
    getPaymentByIdWithHttpInfo(txnId, _options) {
        const result = this.api.getPaymentByIdWithHttpInfo(txnId, _options);
        return result.toPromise();
    }
    /**
     * This will fetch the given payment associated with this transaction id. Single payment transaction IDs begin with `tbx-`
     * Fetch a payment by its transaction ID
     * @param txnId The payment reference to fetch.
     */
    getPaymentById(txnId, _options) {
        const result = this.api.getPaymentById(txnId, _options);
        return result.toPromise();
    }
    /**
     * This will refund the given payment associated with this transaction id.
     * Refund a payment by its transaction ID
     * @param txnId The payment reference to refund.
     */
    refundPaymentByIdWithHttpInfo(txnId, _options) {
        const result = this.api.refundPaymentByIdWithHttpInfo(txnId, _options);
        return result.toPromise();
    }
    /**
     * This will refund the given payment associated with this transaction id.
     * Refund a payment by its transaction ID
     * @param txnId The payment reference to refund.
     */
    refundPaymentById(txnId, _options) {
        const result = this.api.refundPaymentById(txnId, _options);
        return result.toPromise();
    }
}
exports.PromisePaymentsApi = PromisePaymentsApi;
const ObservableAPI_4 = require("./ObservableAPI");
class PromiseRecurringPaymentsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservableRecurringPaymentsApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * This cancels the recurring payment for the reference provided. Recurring payment references start with `tbx-r-`
     * Cancel a recurring payment
     * @param reference The recurring payment reference to cancel.
     */
    cancelRecurringPaymentWithHttpInfo(reference, _options) {
        const result = this.api.cancelRecurringPaymentWithHttpInfo(reference, _options);
        return result.toPromise();
    }
    /**
     * This cancels the recurring payment for the reference provided. Recurring payment references start with `tbx-r-`
     * Cancel a recurring payment
     * @param reference The recurring payment reference to cancel.
     */
    cancelRecurringPayment(reference, _options) {
        const result = this.api.cancelRecurringPayment(reference, _options);
        return result.toPromise();
    }
    /**
     * Fetch a recurring payment (subscription) by its reference
     * @param reference The recurring payment reference to fetch.
     */
    getRecurringPaymentWithHttpInfo(reference, _options) {
        const result = this.api.getRecurringPaymentWithHttpInfo(reference, _options);
        return result.toPromise();
    }
    /**
     * Fetch a recurring payment (subscription) by its reference
     * @param reference The recurring payment reference to fetch.
     */
    getRecurringPayment(reference, _options) {
        const result = this.api.getRecurringPayment(reference, _options);
        return result.toPromise();
    }
    /**
     * Pause or reactivate a recurring payment
     * @param reference The recurring payment reference to update.
     * @param [updateRecurringPaymentRequest]
     */
    updateRecurringPaymentWithHttpInfo(reference, updateRecurringPaymentRequest, _options) {
        const result = this.api.updateRecurringPaymentWithHttpInfo(reference, updateRecurringPaymentRequest, _options);
        return result.toPromise();
    }
    /**
     * Pause or reactivate a recurring payment
     * @param reference The recurring payment reference to update.
     * @param [updateRecurringPaymentRequest]
     */
    updateRecurringPayment(reference, updateRecurringPaymentRequest, _options) {
        const result = this.api.updateRecurringPayment(reference, updateRecurringPaymentRequest, _options);
        return result.toPromise();
    }
    /**
     * If the new subscription amount is higher than the existing amount, a pro-rata charge will be made to cover the cost of the new price up until the next billing date.   **This endpoint requires prior approval - please contact your account manager.**
     * Update a subscription with a new product / amount to pay - replacing the existing product
     * @param reference The recurring payment reference to fetch.
     * @param [updateSubscriptionRequest]
     */
    updateSubscriptionWithHttpInfo(reference, updateSubscriptionRequest, _options) {
        const result = this.api.updateSubscriptionWithHttpInfo(reference, updateSubscriptionRequest, _options);
        return result.toPromise();
    }
    /**
     * If the new subscription amount is higher than the existing amount, a pro-rata charge will be made to cover the cost of the new price up until the next billing date.   **This endpoint requires prior approval - please contact your account manager.**
     * Update a subscription with a new product / amount to pay - replacing the existing product
     * @param reference The recurring payment reference to fetch.
     * @param [updateSubscriptionRequest]
     */
    updateSubscription(reference, updateSubscriptionRequest, _options) {
        const result = this.api.updateSubscription(reference, updateSubscriptionRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseRecurringPaymentsApi = PromiseRecurringPaymentsApi;
