// @tebexio/tebex-sdk-nodejs/dist/Headless/headless.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Headless = void 0;
const lib_1 = require("./lib");
//import {BasketFacade} from "./Headless/basketFacade";
class Headless {
    constructor(publicToken) {
        this._publicToken = "";
        this._publicToken = publicToken;
        this._webstore = new lib_1.Webstore();
        this.headlessApi = new lib_1.HeadlessApi((0, lib_1.createConfiguration)());
    }
    createBasket(basketData) {
        return this.headlessApi.createBasket(this._publicToken, basketData);
    }
    getUserAuthUrl(basket, returnUrl) {
        //TODO
    }
    getUsernameFieldId() {
        //TODO return based on webstore type
        return "";
    }
    getCheckoutLinks(basket) {
        return basket.links;
    }
    getBasketByIdent(ident) {
        return this.headlessApi.getBasketById(this._publicToken, ident);
    }
    getAllCategories() {
        return this.headlessApi.getAllCategories(this._publicToken);
    }
    getAllPackages() {
        return this.headlessApi.getAllPackages(this._publicToken);
    }
    getCategory(categoryId) {
        return this.headlessApi.getCategoryById(this._publicToken, categoryId.toString());
    }
    getPackage(packageId) {
        return this.headlessApi.getPackageById(this._publicToken, packageId);
    }
    applyCoupon(basketFacade, coupon) {
        return this.headlessApi.applyCoupon(this._publicToken, basketFacade.getIdent());
    }
    removeCoupon(basketFacade, coupon) {
        return this.headlessApi.removeCoupon(this._publicToken, basketFacade.getIdent());
    }
}
exports.Headless = Headless;
