// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/models/BasketPackage.js
"use strict";
/**
 * Tebex Headless API
 * The headless API is designed for implementing your own store frontend with the data of your store. You are able to call the Headless API directly from a web browser (such as within an SPA), or from a backend server, such as for in-game GUIs.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: tebex-integrations@overwolf.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasketPackageTypeEnum = exports.BasketPackage = void 0;
/**
* A package within an existing basket.
*/
class BasketPackage {
    static getAttributeTypeMap() {
        return BasketPackage.attributeTypeMap;
    }
    constructor() {
    }
}
exports.BasketPackage = BasketPackage;
BasketPackage.discriminator = undefined;
BasketPackage.mapping = undefined;
BasketPackage.attributeTypeMap = [
    {
        "name": "qty",
        "baseName": "qty",
        "type": "number",
        "format": ""
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "BasketPackageTypeEnum",
        "format": ""
    },
    {
        "name": "revenueShare",
        "baseName": "revenue_share",
        "type": "Array<RevenueShare>",
        "format": ""
    }
];
var BasketPackageTypeEnum;
(function (BasketPackageTypeEnum) {
    BasketPackageTypeEnum["Single"] = "single";
    BasketPackageTypeEnum["Subscription"] = "subscription";
})(BasketPackageTypeEnum || (exports.BasketPackageTypeEnum = BasketPackageTypeEnum = {}));
