// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/models/Category.js
"use strict";
/**
 * Tebex Headless API
 * The headless API is designed for implementing your own store frontend with the data of your store. You are able to call the Headless API directly from a web browser (such as within an SPA), or from a backend server, such as for in-game GUIs.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: tebex-integrations@overwolf.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CategoryDisplayTypeEnum = exports.Category = void 0;
class Category {
    static getAttributeTypeMap() {
        return Category.attributeTypeMap;
    }
    constructor() {
    }
}
exports.Category = Category;
Category.discriminator = undefined;
Category.mapping = undefined;
Category.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "number",
        "format": ""
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "string",
        "format": ""
    },
    {
        "name": "slug",
        "baseName": "slug",
        "type": "string",
        "format": ""
    },
    {
        "name": "parent",
        "baseName": "parent",
        "type": "any",
        "format": "Category"
    },
    {
        "name": "description",
        "baseName": "description",
        "type": "string",
        "format": ""
    },
    {
        "name": "packages",
        "baseName": "packages",
        "type": "Array<Package>",
        "format": ""
    },
    {
        "name": "order",
        "baseName": "order",
        "type": "number",
        "format": ""
    },
    {
        "name": "displayType",
        "baseName": "display_type",
        "type": "CategoryDisplayTypeEnum",
        "format": ""
    }
];
var CategoryDisplayTypeEnum;
(function (CategoryDisplayTypeEnum) {
    CategoryDisplayTypeEnum["List"] = "list";
    CategoryDisplayTypeEnum["Grid"] = "grid";
})(CategoryDisplayTypeEnum || (exports.CategoryDisplayTypeEnum = CategoryDisplayTypeEnum = {}));
