// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("../models/AddBasketPackageRequest"), exports);
__exportStar(require("../models/ApplyCreatorCodeRequest"), exports);
__exportStar(require("../models/Basket"), exports);
__exportStar(require("../models/BasketAuthResponseInner"), exports);
__exportStar(require("../models/BasketLinks"), exports);
__exportStar(require("../models/BasketPackage"), exports);
__exportStar(require("../models/BasketResponse"), exports);
__exportStar(require("../models/CMSPage"), exports);
__exportStar(require("../models/CMSPagesResponse"), exports);
__exportStar(require("../models/Category"), exports);
__exportStar(require("../models/CategoryResponse"), exports);
__exportStar(require("../models/Coupon"), exports);
__exportStar(require("../models/CreateBasketRequest"), exports);
__exportStar(require("../models/ErrorResponse"), exports);
__exportStar(require("../models/GiftCard"), exports);
__exportStar(require("../models/Package"), exports);
__exportStar(require("../models/PackageCategory"), exports);
__exportStar(require("../models/PackageResponse"), exports);
__exportStar(require("../models/RemoveBasketPackageRequest"), exports);
__exportStar(require("../models/RemoveGiftCardRequest"), exports);
__exportStar(require("../models/RevenueShare"), exports);
__exportStar(require("../models/UpdatePackageQuantityRequest"), exports);
__exportStar(require("../models/UpdateTierRequest"), exports);
__exportStar(require("../models/UpdateTierResponse"), exports);
__exportStar(require("../models/Webstore"), exports);
__exportStar(require("../models/WebstoreResponse"), exports);
__exportStar(require("../models/WebstoreResponseData"), exports);
const AddBasketPackageRequest_1 = require("../models/AddBasketPackageRequest");
const ApplyCreatorCodeRequest_1 = require("../models/ApplyCreatorCodeRequest");
const Basket_1 = require("../models/Basket");
const BasketAuthResponseInner_1 = require("../models/BasketAuthResponseInner");
const BasketLinks_1 = require("../models/BasketLinks");
const BasketPackage_1 = require("../models/BasketPackage");
const BasketResponse_1 = require("../models/BasketResponse");
const CMSPage_1 = require("../models/CMSPage");
const CMSPagesResponse_1 = require("../models/CMSPagesResponse");
const Category_1 = require("../models/Category");
const CategoryResponse_1 = require("../models/CategoryResponse");
const Coupon_1 = require("../models/Coupon");
const CreateBasketRequest_1 = require("../models/CreateBasketRequest");
const ErrorResponse_1 = require("../models/ErrorResponse");
const GiftCard_1 = require("../models/GiftCard");
const Package_1 = require("../models/Package");
const PackageCategory_1 = require("../models/PackageCategory");
const PackageResponse_1 = require("../models/PackageResponse");
const RemoveBasketPackageRequest_1 = require("../models/RemoveBasketPackageRequest");
const RemoveGiftCardRequest_1 = require("../models/RemoveGiftCardRequest");
const RevenueShare_1 = require("../models/RevenueShare");
const UpdatePackageQuantityRequest_1 = require("../models/UpdatePackageQuantityRequest");
const UpdateTierRequest_1 = require("../models/UpdateTierRequest");
const UpdateTierResponse_1 = require("../models/UpdateTierResponse");
const Webstore_1 = require("../models/Webstore");
const WebstoreResponse_1 = require("../models/WebstoreResponse");
const WebstoreResponseData_1 = require("../models/WebstoreResponseData");
/* tslint:disable:no-unused-variable */
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
let enumsMap = new Set([
    "BasketPackageTypeEnum",
    "CategoryDisplayTypeEnum",
]);
let typeMap = {
    "AddBasketPackageRequest": AddBasketPackageRequest_1.AddBasketPackageRequest,
    "ApplyCreatorCodeRequest": ApplyCreatorCodeRequest_1.ApplyCreatorCodeRequest,
    "Basket": Basket_1.Basket,
    "BasketAuthResponseInner": BasketAuthResponseInner_1.BasketAuthResponseInner,
    "BasketLinks": BasketLinks_1.BasketLinks,
    "BasketPackage": BasketPackage_1.BasketPackage,
    "BasketResponse": BasketResponse_1.BasketResponse,
    "CMSPage": CMSPage_1.CMSPage,
    "CMSPagesResponse": CMSPagesResponse_1.CMSPagesResponse,
    "Category": Category_1.Category,
    "CategoryResponse": CategoryResponse_1.CategoryResponse,
    "Coupon": Coupon_1.Coupon,
    "CreateBasketRequest": CreateBasketRequest_1.CreateBasketRequest,
    "ErrorResponse": ErrorResponse_1.ErrorResponse,
    "GiftCard": GiftCard_1.GiftCard,
    "Package": Package_1.Package,
    "PackageCategory": PackageCategory_1.PackageCategory,
    "PackageResponse": PackageResponse_1.PackageResponse,
    "RemoveBasketPackageRequest": RemoveBasketPackageRequest_1.RemoveBasketPackageRequest,
    "RemoveGiftCardRequest": RemoveGiftCardRequest_1.RemoveGiftCardRequest,
    "RevenueShare": RevenueShare_1.RevenueShare,
    "UpdatePackageQuantityRequest": UpdatePackageQuantityRequest_1.UpdatePackageQuantityRequest,
    "UpdateTierRequest": UpdateTierRequest_1.UpdateTierRequest,
    "UpdateTierResponse": UpdateTierResponse_1.UpdateTierResponse,
    "Webstore": Webstore_1.Webstore,
    "WebstoreResponse": WebstoreResponse_1.WebstoreResponse,
    "WebstoreResponseData": WebstoreResponseData_1.WebstoreResponseData,
};
/**
 * Every mime-type consists of a type, subtype, and optional parameters.
 * The subtype can be composite, including information about the content format.
 * For example: `application/json-patch+json`, `application/merge-patch+json`.
 *
 * This helper transforms a string mime-type into an internal representation.
 * This simplifies the implementation of predicates that in turn define common rules for parsing or stringifying
 * the payload.
 */
const parseMimeType = (mimeType) => {
    const [type = '', subtype = ''] = mimeType.split('/');
    return {
        type,
        subtype,
        subtypeTokens: subtype.split('+'),
    };
};
// This factory creates a predicate function that checks a string mime-type against defined rules.
const mimeTypePredicateFactory = (predicate) => (mimeType) => predicate(parseMimeType(mimeType));
// Use this factory when you need to define a simple predicate based only on type and, if applicable, subtype.
const mimeTypeSimplePredicateFactory = (type, subtype) => mimeTypePredicateFactory((descriptor) => {
    if (descriptor.type !== type)
        return false;
    if (subtype != null && descriptor.subtype !== subtype)
        return false;
    return true;
});
// Creating a set of named predicates that will help us determine how to handle different mime-types
const isTextLikeMimeType = mimeTypeSimplePredicateFactory('text');
const isJsonMimeType = mimeTypeSimplePredicateFactory('application', 'json');
const isJsonLikeMimeType = mimeTypePredicateFactory((descriptor) => descriptor.type === 'application' && descriptor.subtypeTokens.some((item) => item === 'json'));
const isOctetStreamMimeType = mimeTypeSimplePredicateFactory('application', 'octet-stream');
const isFormUrlencodedMimeType = mimeTypeSimplePredicateFactory('application', 'x-www-form-urlencoded');
// Defining a list of mime-types in the order of prioritization for handling.
const supportedMimeTypePredicatesWithPriority = [
    isJsonMimeType,
    isJsonLikeMimeType,
    isTextLikeMimeType,
    isOctetStreamMimeType,
    isFormUrlencodedMimeType,
];
const nullableSuffix = " | null";
const optionalSuffix = " | undefined";
const arrayPrefix = "Array<";
const arraySuffix = ">";
const mapPrefix = "{ [key: string]: ";
const mapSuffix = "; }";
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType; // w/e we don't know the type
            }
            // Check the discriminator
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType; // the type does not have a discriminator. use it.
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    let mapping = typeMap[expectedType].mapping;
                    if (mapping != undefined && mapping[discriminatorType]) {
                        return mapping[discriminatorType]; // use the type given in the discriminator
                    }
                    else if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType; // discriminator did not map to a type
                    }
                }
                else {
                    return expectedType; // discriminator was not present (or an empty string)
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.endsWith(nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.serialize(data, subType, format);
        }
        else if (type.endsWith(optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.serialize(data, subType, format);
        }
        else if (type.startsWith(arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let date of data) {
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type.startsWith(mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.serialize(data[key], subType, format);
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) { // in case we dont know the type
                return data;
            }
            // Get the actual type of this object
            type = this.findCorrectType(data, type);
            // get the map for the correct type.
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let attributeType of attributeTypes) {
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        // polymorphism may change the actual type.
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.endsWith(nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.deserialize(data, subType, format);
        }
        else if (type.endsWith(optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.deserialize(data, subType, format);
        }
        else if (type.startsWith(arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let date of data) {
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type.startsWith(mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.deserialize(data[key], subType, format);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) { // is Enum
                return data;
            }
            if (!typeMap[type]) { // dont know the type
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let attributeType of attributeTypes) {
                let value = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
                if (value !== undefined) {
                    instance[attributeType.name] = value;
                }
            }
            return instance;
        }
    }
    /**
     * Normalize media type
     *
     * We currently do not handle any media types attributes, i.e. anything
     * after a semicolon. All content is assumed to be UTF-8 compatible.
     */
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return (mediaType.split(";")[0] ?? '').trim().toLowerCase();
    }
    /**
     * From a list of possible media types, choose the one we can handle best.
     *
     * The order of the given media types does not have any impact on the choice
     * made.
     */
    static getPreferredMediaType(mediaTypes) {
        /** According to OAS 3 we should default to json */
        if (mediaTypes.length === 0) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        for (const predicate of supportedMimeTypePredicatesWithPriority) {
            for (const mediaType of normalMediaTypes) {
                if (mediaType != null && predicate(mediaType)) {
                    return mediaType;
                }
            }
        }
        throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
    }
    /**
     * Convert data to a string according the given media type
     */
    static stringify(data, mediaType) {
        if (isTextLikeMimeType(mediaType)) {
            return String(data);
        }
        if (isJsonLikeMimeType(mediaType)) {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    /**
     * Parse data from a string according to the given media type
     */
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (isTextLikeMimeType(mediaType)) {
            return rawData;
        }
        if (isJsonLikeMimeType(mediaType)) {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
