// @tebexio/tebex-sdk-nodejs/dist/checkout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Checkout = void 0;
const basketBuilder_1 = require("./Checkout/basketBuilder");
const packageBuilder_1 = require("./Checkout/packageBuilder");
const checkoutApi_1 = require("./Checkout/lib/api/checkoutApi");
const paymentsApi_1 = require("./Checkout/lib/api/paymentsApi");
const recurringPaymentsApi_1 = require("./Checkout/lib/api/recurringPaymentsApi");
const basketsApi_1 = require("./Checkout/lib/api/basketsApi");
const package_1 = require("./Checkout/lib/model/package");
const addPackageRequest_1 = require("./Checkout/lib/model/addPackageRequest");
const updateSubscriptionRequestItemsInner_1 = require("./Checkout/lib/model/updateSubscriptionRequestItemsInner");
const updateRecurringPaymentRequest_1 = require("./Checkout/lib/model/updateRecurringPaymentRequest");
/**
 * TebexCheckout allows creating and transacting with ad-hoc packages that are not pre-defined in a Tebex project.
 *
 * NOTE: This API requires prior approval. Please contact Tebex to enable on your account.
 *
 * This provides an interface to the underlying OpenAPI implementation. If your desired functionality is not here, you may
 * retrieve the OpenAPI objects for manual API interaction.
 */
class Checkout {
    get checkoutApi() {
        return this._checkoutApi;
    }
    get paymentsApi() {
        return this._paymentsApi;
    }
    get recurringPaymentsApi() {
        return this._recurringPaymentsApi;
    }
    get basketsApi() {
        return this._basketsApi;
    }
    constructor() {
        this._areApiKeysSet = false;
        this._checkoutApi = new checkoutApi_1.CheckoutApi("", "");
        this._paymentsApi = new paymentsApi_1.PaymentsApi("", "");
        this._recurringPaymentsApi = new recurringPaymentsApi_1.RecurringPaymentsApi("", "");
        this._basketsApi = new basketsApi_1.BasketsApi("", "");
    }
    areApiKeysSet() {
        return this._areApiKeysSet;
    }
    /**
     * Sets the API keys for the Checkout instance and initializes the required API services.
     *
     * @param {string} projectId - The unique identifier for the project.
     * @param {string} privateKey - The private key associated with the project for authentication.
     * @return {Checkout} The current instance of the Checkout object.
     */
    setApiKeys(projectId, privateKey) {
        this._checkoutApi = new checkoutApi_1.CheckoutApi(projectId, privateKey);
        this._paymentsApi = new paymentsApi_1.PaymentsApi(projectId, privateKey);
        this._recurringPaymentsApi = new recurringPaymentsApi_1.RecurringPaymentsApi(projectId, privateKey);
        this._basketsApi = new basketsApi_1.BasketsApi(projectId, privateKey);
        this._areApiKeysSet = true;
        return this;
    }
    /**
     * Creates and returns a new instance of BasketBuilder.
     *
     * @return {BasketBuilder} A new instance of BasketBuilder to construct and customize a basket.
     */
    newBasketBuilder() {
        return new basketBuilder_1.BasketBuilder();
    }
    /**
     * Creates and returns a new instance of the PackageBuilder.
     *
     * @return {PackageBuilder} A new instance of PackageBuilder.
     */
    newPackageBuilder() {
        return new packageBuilder_1.PackageBuilder();
    }
    /**
     * Creates a new basket using the provided create request. Use the BasketBuilder to make a CreateBasketRequest.
     *
     * @param {CreateBasketRequest} createRequest - The request object containing details required to create a basket.
     * @return {Promise<Basket>} A promise that resolves to the created basket object.
     */
    async createBasket(createRequest) {
        const { response, body } = await this._basketsApi.createBasket(createRequest);
        return body;
    }
    /**
     * Adds a package to the specified basket.
     *
     * @param {Basket} basket - The basket to which the package will be added.
     * @param {Package} pack - The package to add, including its type and quantity.
     * @return {Promise<Basket>} A promise that resolves to the updated basket.
     */
    async addPackage(basket, pack) {
        const { response, body } = await this._basketsApi.addPackage(basket.ident, {
            _package: pack,
            qty: pack.qty,
            type: pack.type == package_1.Package.TypeEnum.Single ? addPackageRequest_1.AddPackageRequest.TypeEnum.Single : addPackageRequest_1.AddPackageRequest.TypeEnum.Subscription
        });
        return body;
    }
    /**
     * Processes a checkout request.
     *
     * @param {BasketBuilder} basketBuilder - The builder used to construct the checkout request payload.
     * @param {any[]} checkoutItems - The items to be included in the checkout request.
     * @param {Sale|undefined} sale - Optional sale details to include in the checkout request. Defaults to undefined.
     * @return {Promise<Basket>} A promise that resolves to the resulting Basket object returned by the API.
     */
    async checkoutRequest(basketBuilder, checkoutItems, sale = undefined) {
        const { response, body } = await this._checkoutApi.checkout({
            basket: basketBuilder.buildCheckoutRequest(),
            items: checkoutItems,
            sale: sale
        });
        return body;
    }
    /**
     * Retrieves a basket by its identifier.
     *
     * @param {string} ident - The unique identifier of the basket to retrieve.
     * @return {Promise<Basket>} A promise that resolves to the retrieved basket object.
     */
    async getBasket(ident) {
        const { response, body } = await this._basketsApi.getBasketById(ident);
        return body;
    }
    /**
     * Removes a specific row from the provided basket.
     *
     * @param {Basket} basket - The basket object from which the row needs to be removed.
     * @param {number} rowId - The unique identifier of the row to remove.
     * @return {Promise<any>} A promise resolving to the response body of the removal operation.
     */
    async removeBasketRow(basket, rowId) {
        const { response, body } = await this._basketsApi.removeRowFromBasket(basket.ident, rowId);
        return body;
    }
    /**
     * Adds a sale to the given basket with the specified details.
     *
     * @param {Basket} basket - The basket to which the sale will be added.
     * @param {string} saleName - The name of the sale.
     * @param {Sale.DiscountTypeEnum} saleType - The type of discount being applied.
     * @param {number} saleAmount - The amount or value of the discount.
     * @return {Promise<Basket>} A promise that resolves to the updated basket with the applied sale.
     */
    async addSaleToBasket(basket, saleName, saleType, saleAmount) {
        const { response, body } = await this._basketsApi.addSaleToBasket(basket.ident, {
            name: saleName,
            discountType: saleType,
            amount: saleAmount,
        });
        return body;
    }
    /**
     * Retrieves payment details using the specified identifier.
     *
     * @param {string} identifier - The unique identifier of the payment to retrieve.
     * @return {Promise<Payment>} A promise that resolves to the payment details.
     */
    async getPayment(identifier) {
        const { response, body } = await this._paymentsApi.getPaymentById(identifier);
        return body;
    }
    /**
     * Processes a refund for a given payment transaction.
     *
     * @param {string} transactionId - The unique identifier of the transaction to be refunded.
     * @return {Promise<Payment>} A promise that resolves to the refunded payment details.
     */
    async refundPayment(transactionId) {
        const { response, body } = await this._paymentsApi.refundPaymentById(transactionId);
        return body;
    }
    /**
     * Retrieves a recurring payment using the specified transaction ID.
     *
     * @param {string} transactionId - The unique identifier for the transaction
     *                                  associated with the recurring payment.
     * @return {Promise<RecurringPayment>} A promise that resolves to the details
     *                                      of the recurring payment.
     */
    async getRecurringPayment(transactionId) {
        const { response, body } = await this._recurringPaymentsApi.getRecurringPayment(transactionId);
        return body;
    }
    /**
     * Updates the subscription product associated with the given recurring payment reference.
     *
     * @param {string} recurringPaymentReference - The unique reference identifier for the recurring payment.
     * @param {Package[]} packages - An array of packages to be updated in the subscription. Each package includes details about the product, quantity, and type.
     * @return {Promise<RecurringPayment>} A promise resolving to the updated recurring payment object.
     */
    async updateSubscriptionProduct(recurringPaymentReference, packages) {
        let items = [];
        packages.forEach(p => {
            items.push({
                type: p.type == package_1.Package.TypeEnum.Single ? updateSubscriptionRequestItemsInner_1.UpdateSubscriptionRequestItemsInner.TypeEnum.Single : updateSubscriptionRequestItemsInner_1.UpdateSubscriptionRequestItemsInner.TypeEnum.Subscription,
                qty: p.qty,
                _package: p,
            });
        });
        const { response, body } = await this._recurringPaymentsApi.updateSubscription(recurringPaymentReference, {
            items: items
        });
        return body;
    }
    /**
     * Cancels a recurring payment using the provided recurring payment reference.
     *
     * @param {string} recurringPaymentReference - The unique identifier for the recurring payment to be canceled.
     * @return {Promise<RecurringPayment>} A promise that resolves to the details of the canceled recurring payment.
     */
    async cancelRecurringPayment(recurringPaymentReference) {
        const { response, body } = await this._recurringPaymentsApi.cancelRecurringPayment(recurringPaymentReference);
        return body;
    }
    /**
     * Pauses a recurring payment by updating the payment's status to "Paused".
     *
     * @param {string} recurringPaymentReference - The unique reference identifier for the recurring payment to be paused.
     * @return {Promise<RecurringPayment>} A promise that resolves to the updated RecurringPayment object.
     */
    async pauseRecurringPayment(recurringPaymentReference) {
        const { response, body } = await this._recurringPaymentsApi.updateRecurringPayment(recurringPaymentReference, {
            status: updateRecurringPaymentRequest_1.UpdateRecurringPaymentRequest.StatusEnum.Paused
        });
        return body;
    }
    /**
     * Reactivates a recurring payment that is currently inactive.
     *
     * @param {string} recurringPaymentReference - The unique reference identifier for the recurring payment to be reactivated.
     * @return {Promise<RecurringPayment>} A promise that resolves with the reactivated recurring payment details.
     */
    async reactivateRecurringPayment(recurringPaymentReference) {
        const { response, body } = await this._recurringPaymentsApi.updateRecurringPayment(recurringPaymentReference, {
            status: updateRecurringPaymentRequest_1.UpdateRecurringPaymentRequest.StatusEnum.Active
        });
        return body;
    }
}
exports.Checkout = Checkout;
