// @tebexio/tebex-sdk-nodejs/dist/headless.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Headless = void 0;
const headlessApi_1 = require("./Headless/lib/api/headlessApi");
const webstore_1 = require("./Headless/lib/model/webstore");
const tebexProject_1 = require("./Headless/projects/tebexProject");
/**
 * TebexHeadless allows you to access and create baskets for your Tebex project using pre-defined packages.
 */
class Headless {
    constructor() {
        this._publicToken = "";
        this._headlessApi = new headlessApi_1.HeadlessApi();
        this._webstore = new webstore_1.Webstore();
    }
    getHeadlessApi() {
        return this._headlessApi;
    }
    /**
     * Sets the current project using the provided public token and retrieves relevant project data.
     *
     * @param {string} publicToken - The public token used to identify and authenticate the project.
     * @return {Promise<TebexProject>} A promise that resolves to an instance of the TebexProject.
     */
    async setProject(publicToken) {
        this._publicToken = publicToken;
        this._headlessApi = new headlessApi_1.HeadlessApi("https://headless.tebex.io/api");
        return this._headlessApi.getWebstoreById(this._publicToken).then(({ body }) => {
            this._webstore = body.data?.schema;
            return new tebexProject_1.TebexProject(this._webstore);
        });
    }
    /**
     * Creates a new basket using the provided basket data.
     *
     * @param {CreateBasketRequest} basketData - The data required to create a new basket.
     * @return {Promise<Basket>} A promise that resolves to the created basket.
     */
    async createBasket(basketData) {
        const { body } = await this._headlessApi.createBasket(this._publicToken, basketData);
        return body.data;
    }
    /**
     * Retrieves the authentication URL for the specified basket.
     *
     * @param {Basket} basket - The basket object containing the necessary identification information.
     * @param {string} returnUrl - The URL to which the user is redirected after authentication.
     * @return {Promise<string>} A promise resolving to the authentication URL as a string, or an empty string if the URL cannot be retrieved.
     */
    async getUserAuthUrl(basket, returnUrl) {
        return this._headlessApi.getBasketAuthUrl(this._publicToken, basket.ident, returnUrl).then(({ body }) => {
            if (body.length == 0) {
                return "";
            }
            return body[0].url;
        });
    }
    /**
     * Retrieves the checkout-related links from the provided basket object.
     *
     * @param {Basket} basket - The basket object containing checkout-related links.
     * @return {BasketLinks} The collection of checkout-related links associated with the basket.
     */
    getCheckoutLinks(basket) {
        return basket.links;
    }
    /**
     * Retrieves a basket by its identifier.
     *
     * @param {string} ident - The unique identifier for the basket.
     * @return {Promise<Basket>} A promise that resolves to the basket data associated with the given identifier.
     */
    async getBasketByIdent(ident) {
        const { body } = await this._headlessApi.getBasketById(this._publicToken, ident);
        return body.data;
    }
    /**
     * Retrieves all categories available from the API.
     *
     * @return {Promise<Category[]>} A promise that resolves to an array of Category objects.
     */
    async getAllCategories() {
        const { body } = await this._headlessApi.getAllCategories(this._publicToken);
        return body.data;
    }
    /**
     * Retrieves all packages available via the headless API.
     *
     * @return {Promise<Package[]>} A promise that resolves to an array of Package objects.
     */
    async getAllPackages() {
        const { body } = await this._headlessApi.getAllPackages(this._publicToken);
        return body.data;
    }
    /**
     * Retrieves the category information for the specified category ID.
     *
     * @param {number} categoryId - The unique identifier of the category to retrieve.
     * @return {Promise<Category>} A Promise that resolves to the category data.
     */
    async getCategory(categoryId) {
        const { body } = await this._headlessApi.getCategoryById(this._publicToken, categoryId.toString());
        return body.data[0];
    }
    /**
     * Retrieves the package details based on the provided package ID.
     *
     * @param {number} packageId - The unique identifier of the package to retrieve.
     * @return {Promise<Package>} A promise that resolves to the package details.
     */
    async getPackage(packageId) {
        const { body } = await this._headlessApi.getPackageById(this._publicToken, packageId);
        return body.data[0];
    }
    /**
     * Applies a coupon to the given basket using the provided basket facade.
     *
     * @param {BasketFacade} basketFacade - The facade representing the basket to which the coupon will be applied.
     * @param {Coupon} coupon - The coupon to be applied to the basket.
     * @return {Promise<Basket>} A promise resolving to the updated basket after the coupon is applied.
     */
    async applyCoupon(basketFacade, coupon) {
        const { body } = await this._headlessApi.applyCoupon(this._publicToken, basketFacade.getIdent(), coupon);
        return body.data;
    }
    /**
     * Removes a coupon from the basket.
     *
     * @param {BasketFacade} basketFacade - The basket facade instance that provides the identifier of the basket from which the coupon should be removed.
     * @return {Promise<{ response: http.IncomingMessage; body?: any }>} A promise that resolves with the HTTP response and optionally the response body.
     */
    async removeCoupon(basketFacade) {
        return this._headlessApi.removeCoupon(this._publicToken, basketFacade.getIdent());
    }
    /**
     * Adds a creator code to the specified basket.
     *
     * @param {BasketFacade} basketFacade - The facade object representing the basket.
     * @param {string} code - The creator code to be added to the basket.
     * @return {Promise<Basket>} A promise that resolves to the updated basket.
     */
    async addCreatorCode(basketFacade, code) {
        const { body } = await this._headlessApi.applyCreatorCode(this._publicToken, basketFacade.getIdent(), {
            creatorCode: code
        });
        return body.data;
    }
    /**
     * Removes the creator code associated with the provided BasketFacade instance.
     *
     * @param {BasketFacade} basketFacade - The instance of BasketFacade whose creator code is to be removed.
     * @return {Promise<Basket>} A promise that resolves with the updated Basket after the creator code has been removed.
     */
    async removeCreatorCode(basketFacade) {
        const { body } = await this._headlessApi.removeCreatorCode(this._publicToken, basketFacade.getIdent());
        return body.data;
    }
    /**
     * Adds a gift card to the provided basket.
     *
     * @param {BasketFacade} basketFacade - The basket facade instance representing the current basket.
     * @param {string} cardNumber - The gift card number to be applied to the basket.
     * @return {Promise<Basket>} A promise that resolves to the updated basket after the gift card has been applied.
     */
    async addGiftCard(basketFacade, cardNumber) {
        const { body } = await this._headlessApi.applyGiftCard(this._publicToken, basketFacade.getIdent(), {
            cardNumber: cardNumber
        });
        return body.data;
    }
    /**
     * Removes a gift card from the specified basket.
     *
     * @param basketFacade - The basketFacade object that provides access to the basket's identity and details.
     * @param cardNumber - The unique number of the gift card to be removed.
     * @return A promise that resolves to the updated basket object.
     */
    async removeGiftCard(basketFacade, cardNumber) {
        const { body } = await this._headlessApi.removeGiftCard(this._publicToken, basketFacade.getIdent(), {
            cardNumber: cardNumber
        });
        return body.data;
    }
    /**
     * Retrieves the webstore instance.
     *
     * @return {Object} The webstore object associated with this instance.
     */
    getWebstore() {
        return this._webstore;
    }
}
exports.Headless = Headless;
