// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/api/apis.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIS = exports.HttpError = void 0;
__exportStar(require("./basketsApi"), exports);
const basketsApi_1 = require("./basketsApi");
__exportStar(require("./checkoutApi"), exports);
const checkoutApi_1 = require("./checkoutApi");
__exportStar(require("./paymentsApi"), exports);
const paymentsApi_1 = require("./paymentsApi");
__exportStar(require("./recurringPaymentsApi"), exports);
const recurringPaymentsApi_1 = require("./recurringPaymentsApi");
class HttpError extends Error {
    constructor(response, body, statusCode) {
        super('HTTP request failed');
        this.response = response;
        this.body = body;
        this.statusCode = statusCode;
        this.name = 'HttpError';
    }
}
exports.HttpError = HttpError;
exports.APIS = [basketsApi_1.BasketsApi, checkoutApi_1.CheckoutApi, paymentsApi_1.PaymentsApi, recurringPaymentsApi_1.RecurringPaymentsApi];
