// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/auth/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TebexCheckoutAuthBasicAuthentication = void 0;
exports.configureAuthMethods = configureAuthMethods;
/**
 * Applies http authentication to the request context.
 */
class TebexCheckoutAuthBasicAuthentication {
    /**
     * Configures the http authentication with the required details.
     *
     * @param username username for http basic authentication
     * @param password password for http basic authentication
     */
    constructor(username, password) {
        this.username = username;
        this.password = password;
    }
    getName() {
        return "tebex_checkout_auth_basic";
    }
    applySecurityAuthentication(context) {
        let comb = Buffer.from(this.username + ":" + this.password, 'binary').toString('base64');
        context.setHeaderParam("Authorization", "Basic " + comb);
    }
}
exports.TebexCheckoutAuthBasicAuthentication = TebexCheckoutAuthBasicAuthentication;
/**
 * Creates the authentication methods from a swagger description.
 *
 */
function configureAuthMethods(config) {
    let authMethods = {};
    if (!config) {
        return authMethods;
    }
    authMethods["default"] = config["default"];
    if (config["tebex_checkout_auth_basic"]) {
        authMethods["tebex_checkout_auth_basic"] = new TebexCheckoutAuthBasicAuthentication(config["tebex_checkout_auth_basic"]["username"], config["tebex_checkout_auth_basic"]["password"]);
    }
    return authMethods;
}
