// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/types/ObjectParamAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectHeadlessApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class ObjectHeadlessApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableHeadlessApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * Add a package with the given ID to the basket.
     * Add a package to a basket
     * @param param the request object
     */
    addBasketPackageWithHttpInfo(param, options) {
        return this.api.addBasketPackageWithHttpInfo(param.basketIdent, param.addBasketPackageRequest, options).toPromise();
    }
    /**
     * Add a package with the given ID to the basket.
     * Add a package to a basket
     * @param param the request object
     */
    addBasketPackage(param, options) {
        return this.api.addBasketPackage(param.basketIdent, param.addBasketPackageRequest, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a coupon to a basket.
     * @param param the request object
     */
    applyCouponWithHttpInfo(param, options) {
        return this.api.applyCouponWithHttpInfo(param.token, param.basketIdent, param.coupon, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a coupon to a basket.
     * @param param the request object
     */
    applyCoupon(param, options) {
        return this.api.applyCoupon(param.token, param.basketIdent, param.coupon, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a creator code to a basket.
     * @param param the request object
     */
    applyCreatorCodeWithHttpInfo(param, options) {
        return this.api.applyCreatorCodeWithHttpInfo(param.token, param.basketIdent, param.applyCreatorCodeRequest, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a creator code to a basket.
     * @param param the request object
     */
    applyCreatorCode(param, options) {
        return this.api.applyCreatorCode(param.token, param.basketIdent, param.applyCreatorCodeRequest, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a gift card to a basket.
     * @param param the request object
     */
    applyGiftCardWithHttpInfo(param, options) {
        return this.api.applyGiftCardWithHttpInfo(param.token, param.basketIdent, param.giftCard, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a gift card to a basket.
     * @param param the request object
     */
    applyGiftCard(param, options) {
        return this.api.applyGiftCard(param.token, param.basketIdent, param.giftCard, options).toPromise();
    }
    /**
     * Creates a new basket for use during checkout.
     * Create a new basket
     * @param param the request object
     */
    createBasketWithHttpInfo(param, options) {
        return this.api.createBasketWithHttpInfo(param.token, param.createBasketRequest, options).toPromise();
    }
    /**
     * Creates a new basket for use during checkout.
     * Create a new basket
     * @param param the request object
     */
    createBasket(param, options) {
        return this.api.createBasket(param.token, param.createBasketRequest, options).toPromise();
    }
    /**
     * Gets all categories from a webstore. This does not include package information. To include package information, add `?includePackages=1` to the URL.
     * Gets all categories available in the webstore.
     * @param param the request object
     */
    getAllCategoriesWithHttpInfo(param, options) {
        return this.api.getAllCategoriesWithHttpInfo(param.token, options).toPromise();
    }
    /**
     * Gets all categories from a webstore. This does not include package information. To include package information, add `?includePackages=1` to the URL.
     * Gets all categories available in the webstore.
     * @param param the request object
     */
    getAllCategories(param, options) {
        return this.api.getAllCategories(param.token, options).toPromise();
    }
    /**
     * Gets all categories from a webstore including packages.
     * Gets a store\'s categories including all package information with them.
     * @param param the request object
     */
    getAllCategoriesIncludingPackagesWithHttpInfo(param, options) {
        return this.api.getAllCategoriesIncludingPackagesWithHttpInfo(param.token, options).toPromise();
    }
    /**
     * Gets all categories from a webstore including packages.
     * Gets a store\'s categories including all package information with them.
     * @param param the request object
     */
    getAllCategoriesIncludingPackages(param, options) {
        return this.api.getAllCategoriesIncludingPackages(param.token, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch all packages from a webstore
     * @param param the request object
     */
    getAllPackagesWithHttpInfo(param, options) {
        return this.api.getAllPackagesWithHttpInfo(param.token, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch all packages from a webstore
     * @param param the request object
     */
    getAllPackages(param, options) {
        return this.api.getAllPackages(param.token, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithAuthedIPWithHttpInfo(param, options) {
        return this.api.getAllPackagesWithAuthedIPWithHttpInfo(param.token, param.ipAddress, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithAuthedIP(param, options) {
        return this.api.getAllPackagesWithAuthedIP(param.token, param.ipAddress, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithAuthedIPAndBasketWithHttpInfo(param, options) {
        return this.api.getAllPackagesWithAuthedIPAndBasketWithHttpInfo(param.token, param.basketIdent, param.ipAddress, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithAuthedIPAndBasket(param, options) {
        return this.api.getAllPackagesWithAuthedIPAndBasket(param.token, param.basketIdent, param.ipAddress, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithBasketWithHttpInfo(param, options) {
        return this.api.getAllPackagesWithBasketWithHttpInfo(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getAllPackagesWithBasket(param, options) {
        return this.api.getAllPackagesWithBasket(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Fetches a basket\'s auth URL.
     * Get authentication links for a basket.
     * @param param the request object
     */
    getBasketAuthUrlWithHttpInfo(param, options) {
        return this.api.getBasketAuthUrlWithHttpInfo(param.token, param.basketIdent, param.returnUrl, options).toPromise();
    }
    /**
     * Fetches a basket\'s auth URL.
     * Get authentication links for a basket.
     * @param param the request object
     */
    getBasketAuthUrl(param, options) {
        return this.api.getBasketAuthUrl(param.token, param.basketIdent, param.returnUrl, options).toPromise();
    }
    /**
     * Gets a basket associated with the provided identifier.
     * Fetch a basket from a webstore by its identifier
     * @param param the request object
     */
    getBasketByIdWithHttpInfo(param, options) {
        return this.api.getBasketByIdWithHttpInfo(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Gets a basket associated with the provided identifier.
     * Fetch a basket from a webstore by its identifier
     * @param param the request object
     */
    getBasketById(param, options) {
        return this.api.getBasketById(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Gets a list of custom pages associated with the webstore. These contain a `content` variable with the HTML content of the page.
     * Fetch the custom pages associated with the store.
     * @param param the request object
     */
    getCMSPagesWithHttpInfo(param, options) {
        return this.api.getCMSPagesWithHttpInfo(param.token, options).toPromise();
    }
    /**
     * Gets a list of custom pages associated with the webstore. These contain a `content` variable with the HTML content of the page.
     * Fetch the custom pages associated with the store.
     * @param param the request object
     */
    getCMSPages(param, options) {
        return this.api.getCMSPages(param.token, options).toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category
     * @param param the request object
     */
    getCategoryByIdWithHttpInfo(param, options) {
        return this.api.getCategoryByIdWithHttpInfo(param.token, param.categoryId, options).toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category
     * @param param the request object
     */
    getCategoryById(param, options) {
        return this.api.getCategoryById(param.token, param.categoryId, options).toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category, including all the packages in the category
     * @param param the request object
     */
    getCategoryIncludingPackagesWithHttpInfo(param, options) {
        return this.api.getCategoryIncludingPackagesWithHttpInfo(param.token, param.categoryId, options).toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category, including all the packages in the category
     * @param param the request object
     */
    getCategoryIncludingPackages(param, options) {
        return this.api.getCategoryIncludingPackages(param.token, param.categoryId, options).toPromise();
    }
    /**
     * Gets a package from a webstore by ID.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getPackageByIdWithHttpInfo(param, options) {
        return this.api.getPackageByIdWithHttpInfo(param.token, param.packageId, options).toPromise();
    }
    /**
     * Gets a package from a webstore by ID.
     * Fetch a package from a webstore by its identifier
     * @param param the request object
     */
    getPackageById(param, options) {
        return this.api.getPackageById(param.token, param.packageId, options).toPromise();
    }
    /**
     * Gets the webstore associated with the provided identifier.
     * Fetch a webstore by its identifier
     * @param param the request object
     */
    getWebstoreByIdWithHttpInfo(param, options) {
        return this.api.getWebstoreByIdWithHttpInfo(param.token, options).toPromise();
    }
    /**
     * Gets the webstore associated with the provided identifier.
     * Fetch a webstore by its identifier
     * @param param the request object
     */
    getWebstoreById(param, options) {
        return this.api.getWebstoreById(param.token, options).toPromise();
    }
    /**
     * Remove the given package ID from the basket.
     * Remove a package from a basket
     * @param param the request object
     */
    removeBasketPackageWithHttpInfo(param, options) {
        return this.api.removeBasketPackageWithHttpInfo(param.basketIdent, param.removeBasketPackageRequest, options).toPromise();
    }
    /**
     * Remove the given package ID from the basket.
     * Remove a package from a basket
     * @param param the request object
     */
    removeBasketPackage(param, options) {
        return this.api.removeBasketPackage(param.basketIdent, param.removeBasketPackageRequest, options).toPromise();
    }
    /**
     * Removes a coupon code from a basket
     * Remove a coupon from the basket.
     * @param param the request object
     */
    removeCouponWithHttpInfo(param, options) {
        return this.api.removeCouponWithHttpInfo(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Removes a coupon code from a basket
     * Remove a coupon from the basket.
     * @param param the request object
     */
    removeCoupon(param, options) {
        return this.api.removeCoupon(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Remove a creator code from the basket.
     * @param param the request object
     */
    removeCreatorCodeWithHttpInfo(param, options) {
        return this.api.removeCreatorCodeWithHttpInfo(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Remove a creator code from the basket.
     * @param param the request object
     */
    removeCreatorCode(param, options) {
        return this.api.removeCreatorCode(param.token, param.basketIdent, options).toPromise();
    }
    /**
     * Removes a gift card from the basket.
     * Remove a gift card from the basket.
     * @param param the request object
     */
    removeGiftCardWithHttpInfo(param, options) {
        return this.api.removeGiftCardWithHttpInfo(param.token, param.basketIdent, param.removeGiftCardRequest, options).toPromise();
    }
    /**
     * Removes a gift card from the basket.
     * Remove a gift card from the basket.
     * @param param the request object
     */
    removeGiftCard(param, options) {
        return this.api.removeGiftCard(param.token, param.basketIdent, param.removeGiftCardRequest, options).toPromise();
    }
    /**
     * Sets the quantity of the given item in the basket.
     * Updates the quantity of the given package in the basket. The user must be logged in before the quantity can be changed.
     * @param param the request object
     */
    updatePackageQuantityWithHttpInfo(param, options) {
        return this.api.updatePackageQuantityWithHttpInfo(param.basketIdent, param.packageId, param.updatePackageQuantityRequest, options).toPromise();
    }
    /**
     * Sets the quantity of the given item in the basket.
     * Updates the quantity of the given package in the basket. The user must be logged in before the quantity can be changed.
     * @param param the request object
     */
    updatePackageQuantity(param, options) {
        return this.api.updatePackageQuantity(param.basketIdent, param.packageId, param.updatePackageQuantityRequest, options).toPromise();
    }
    /**
     * Updates a tier.
     * TODO
     * @param param the request object
     */
    updateTierWithHttpInfo(param, options) {
        return this.api.updateTierWithHttpInfo(param.token, param.tierId, param.updateTierRequest, options).toPromise();
    }
    /**
     * Updates a tier.
     * TODO
     * @param param the request object
     */
    updateTier(param, options) {
        return this.api.updateTier(param.token, param.tierId, param.updateTierRequest, options).toPromise();
    }
}
exports.ObjectHeadlessApi = ObjectHeadlessApi;
