// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseHeadlessApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseHeadlessApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableHeadlessApi(configuration, requestFactory, responseProcessor);
    }
    /**
     * Add a package with the given ID to the basket.
     * Add a package to a basket
     * @param basketIdent The basket identifier.
     * @param [addBasketPackageRequest]
     */
    addBasketPackageWithHttpInfo(basketIdent, addBasketPackageRequest, _options) {
        const result = this.api.addBasketPackageWithHttpInfo(basketIdent, addBasketPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * Add a package with the given ID to the basket.
     * Add a package to a basket
     * @param basketIdent The basket identifier.
     * @param [addBasketPackageRequest]
     */
    addBasketPackage(basketIdent, addBasketPackageRequest, _options) {
        const result = this.api.addBasketPackage(basketIdent, addBasketPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a coupon to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [coupon] Provide a &#x60;coupon_code&#x60; to apply to the basket.
     */
    applyCouponWithHttpInfo(token, basketIdent, coupon, _options) {
        const result = this.api.applyCouponWithHttpInfo(token, basketIdent, coupon, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a coupon to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [coupon] Provide a &#x60;coupon_code&#x60; to apply to the basket.
     */
    applyCoupon(token, basketIdent, coupon, _options) {
        const result = this.api.applyCoupon(token, basketIdent, coupon, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a creator code to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [applyCreatorCodeRequest] Provide a &#x60;creator_code&#x60; to apply to the basket.
     */
    applyCreatorCodeWithHttpInfo(token, basketIdent, applyCreatorCodeRequest, _options) {
        const result = this.api.applyCreatorCodeWithHttpInfo(token, basketIdent, applyCreatorCodeRequest, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a creator code to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [applyCreatorCodeRequest] Provide a &#x60;creator_code&#x60; to apply to the basket.
     */
    applyCreatorCode(token, basketIdent, applyCreatorCodeRequest, _options) {
        const result = this.api.applyCreatorCode(token, basketIdent, applyCreatorCodeRequest, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a gift card to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [giftCard] Provide a &#x60;card_number&#x60; to apply to the basket.
     */
    applyGiftCardWithHttpInfo(token, basketIdent, giftCard, _options) {
        const result = this.api.applyGiftCardWithHttpInfo(token, basketIdent, giftCard, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Apply a gift card to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [giftCard] Provide a &#x60;card_number&#x60; to apply to the basket.
     */
    applyGiftCard(token, basketIdent, giftCard, _options) {
        const result = this.api.applyGiftCard(token, basketIdent, giftCard, _options);
        return result.toPromise();
    }
    /**
     * Creates a new basket for use during checkout.
     * Create a new basket
     * @param token The webstore identifier.
     * @param [createBasketRequest]
     */
    createBasketWithHttpInfo(token, createBasketRequest, _options) {
        const result = this.api.createBasketWithHttpInfo(token, createBasketRequest, _options);
        return result.toPromise();
    }
    /**
     * Creates a new basket for use during checkout.
     * Create a new basket
     * @param token The webstore identifier.
     * @param [createBasketRequest]
     */
    createBasket(token, createBasketRequest, _options) {
        const result = this.api.createBasket(token, createBasketRequest, _options);
        return result.toPromise();
    }
    /**
     * Gets all categories from a webstore. This does not include package information. To include package information, add `?includePackages=1` to the URL.
     * Gets all categories available in the webstore.
     * @param token The webstore identifier.
     */
    getAllCategoriesWithHttpInfo(token, _options) {
        const result = this.api.getAllCategoriesWithHttpInfo(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all categories from a webstore. This does not include package information. To include package information, add `?includePackages=1` to the URL.
     * Gets all categories available in the webstore.
     * @param token The webstore identifier.
     */
    getAllCategories(token, _options) {
        const result = this.api.getAllCategories(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all categories from a webstore including packages.
     * Gets a store\'s categories including all package information with them.
     * @param token The webstore identifier.
     */
    getAllCategoriesIncludingPackagesWithHttpInfo(token, _options) {
        const result = this.api.getAllCategoriesIncludingPackagesWithHttpInfo(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all categories from a webstore including packages.
     * Gets a store\'s categories including all package information with them.
     * @param token The webstore identifier.
     */
    getAllCategoriesIncludingPackages(token, _options) {
        const result = this.api.getAllCategoriesIncludingPackages(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch all packages from a webstore
     * @param token The webstore identifier.
     */
    getAllPackagesWithHttpInfo(token, _options) {
        const result = this.api.getAllPackagesWithHttpInfo(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch all packages from a webstore
     * @param token The webstore identifier.
     */
    getAllPackages(token, _options) {
        const result = this.api.getAllPackages(token, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    getAllPackagesWithAuthedIPWithHttpInfo(token, ipAddress, _options) {
        const result = this.api.getAllPackagesWithAuthedIPWithHttpInfo(token, ipAddress, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    getAllPackagesWithAuthedIP(token, ipAddress, _options) {
        const result = this.api.getAllPackagesWithAuthedIP(token, ipAddress, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    getAllPackagesWithAuthedIPAndBasketWithHttpInfo(token, basketIdent, ipAddress, _options) {
        const result = this.api.getAllPackagesWithAuthedIPAndBasketWithHttpInfo(token, basketIdent, ipAddress, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    getAllPackagesWithAuthedIPAndBasket(token, basketIdent, ipAddress, _options) {
        const result = this.api.getAllPackagesWithAuthedIPAndBasket(token, basketIdent, ipAddress, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    getAllPackagesWithBasketWithHttpInfo(token, basketIdent, _options) {
        const result = this.api.getAllPackagesWithBasketWithHttpInfo(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    getAllPackagesWithBasket(token, basketIdent, _options) {
        const result = this.api.getAllPackagesWithBasket(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Fetches a basket\'s auth URL.
     * Get authentication links for a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param returnUrl The URL you would like to redirect the user to after successful basket authentication.
     */
    getBasketAuthUrlWithHttpInfo(token, basketIdent, returnUrl, _options) {
        const result = this.api.getBasketAuthUrlWithHttpInfo(token, basketIdent, returnUrl, _options);
        return result.toPromise();
    }
    /**
     * Fetches a basket\'s auth URL.
     * Get authentication links for a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param returnUrl The URL you would like to redirect the user to after successful basket authentication.
     */
    getBasketAuthUrl(token, basketIdent, returnUrl, _options) {
        const result = this.api.getBasketAuthUrl(token, basketIdent, returnUrl, _options);
        return result.toPromise();
    }
    /**
     * Gets a basket associated with the provided identifier.
     * Fetch a basket from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    getBasketByIdWithHttpInfo(token, basketIdent, _options) {
        const result = this.api.getBasketByIdWithHttpInfo(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Gets a basket associated with the provided identifier.
     * Fetch a basket from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    getBasketById(token, basketIdent, _options) {
        const result = this.api.getBasketById(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Gets a list of custom pages associated with the webstore. These contain a `content` variable with the HTML content of the page.
     * Fetch the custom pages associated with the store.
     * @param token The webstore identifier.
     */
    getCMSPagesWithHttpInfo(token, _options) {
        const result = this.api.getCMSPagesWithHttpInfo(token, _options);
        return result.toPromise();
    }
    /**
     * Gets a list of custom pages associated with the webstore. These contain a `content` variable with the HTML content of the page.
     * Fetch the custom pages associated with the store.
     * @param token The webstore identifier.
     */
    getCMSPages(token, _options) {
        const result = this.api.getCMSPages(token, _options);
        return result.toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    getCategoryByIdWithHttpInfo(token, categoryId, _options) {
        const result = this.api.getCategoryByIdWithHttpInfo(token, categoryId, _options);
        return result.toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    getCategoryById(token, categoryId, _options) {
        const result = this.api.getCategoryById(token, categoryId, _options);
        return result.toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category, including all the packages in the category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    getCategoryIncludingPackagesWithHttpInfo(token, categoryId, _options) {
        const result = this.api.getCategoryIncludingPackagesWithHttpInfo(token, categoryId, _options);
        return result.toPromise();
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category, including all the packages in the category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    getCategoryIncludingPackages(token, categoryId, _options) {
        const result = this.api.getCategoryIncludingPackages(token, categoryId, _options);
        return result.toPromise();
    }
    /**
     * Gets a package from a webstore by ID.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param packageId The package\&#39;s ID.
     */
    getPackageByIdWithHttpInfo(token, packageId, _options) {
        const result = this.api.getPackageByIdWithHttpInfo(token, packageId, _options);
        return result.toPromise();
    }
    /**
     * Gets a package from a webstore by ID.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param packageId The package\&#39;s ID.
     */
    getPackageById(token, packageId, _options) {
        const result = this.api.getPackageById(token, packageId, _options);
        return result.toPromise();
    }
    /**
     * Gets the webstore associated with the provided identifier.
     * Fetch a webstore by its identifier
     * @param token The webstore identifier.
     */
    getWebstoreByIdWithHttpInfo(token, _options) {
        const result = this.api.getWebstoreByIdWithHttpInfo(token, _options);
        return result.toPromise();
    }
    /**
     * Gets the webstore associated with the provided identifier.
     * Fetch a webstore by its identifier
     * @param token The webstore identifier.
     */
    getWebstoreById(token, _options) {
        const result = this.api.getWebstoreById(token, _options);
        return result.toPromise();
    }
    /**
     * Remove the given package ID from the basket.
     * Remove a package from a basket
     * @param basketIdent The basket identifier.
     * @param [removeBasketPackageRequest]
     */
    removeBasketPackageWithHttpInfo(basketIdent, removeBasketPackageRequest, _options) {
        const result = this.api.removeBasketPackageWithHttpInfo(basketIdent, removeBasketPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * Remove the given package ID from the basket.
     * Remove a package from a basket
     * @param basketIdent The basket identifier.
     * @param [removeBasketPackageRequest]
     */
    removeBasketPackage(basketIdent, removeBasketPackageRequest, _options) {
        const result = this.api.removeBasketPackage(basketIdent, removeBasketPackageRequest, _options);
        return result.toPromise();
    }
    /**
     * Removes a coupon code from a basket
     * Remove a coupon from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    removeCouponWithHttpInfo(token, basketIdent, _options) {
        const result = this.api.removeCouponWithHttpInfo(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Removes a coupon code from a basket
     * Remove a coupon from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    removeCoupon(token, basketIdent, _options) {
        const result = this.api.removeCoupon(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Remove a creator code from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    removeCreatorCodeWithHttpInfo(token, basketIdent, _options) {
        const result = this.api.removeCreatorCodeWithHttpInfo(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Applies a creator code to a basket.
     * Remove a creator code from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    removeCreatorCode(token, basketIdent, _options) {
        const result = this.api.removeCreatorCode(token, basketIdent, _options);
        return result.toPromise();
    }
    /**
     * Removes a gift card from the basket.
     * Remove a gift card from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [removeGiftCardRequest] Provide the &#x60;card_number&#x60; to remove from the basket.
     */
    removeGiftCardWithHttpInfo(token, basketIdent, removeGiftCardRequest, _options) {
        const result = this.api.removeGiftCardWithHttpInfo(token, basketIdent, removeGiftCardRequest, _options);
        return result.toPromise();
    }
    /**
     * Removes a gift card from the basket.
     * Remove a gift card from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param [removeGiftCardRequest] Provide the &#x60;card_number&#x60; to remove from the basket.
     */
    removeGiftCard(token, basketIdent, removeGiftCardRequest, _options) {
        const result = this.api.removeGiftCard(token, basketIdent, removeGiftCardRequest, _options);
        return result.toPromise();
    }
    /**
     * Sets the quantity of the given item in the basket.
     * Updates the quantity of the given package in the basket. The user must be logged in before the quantity can be changed.
     * @param basketIdent The basket identifier.
     * @param packageId The package identifier.
     * @param [updatePackageQuantityRequest]
     */
    updatePackageQuantityWithHttpInfo(basketIdent, packageId, updatePackageQuantityRequest, _options) {
        const result = this.api.updatePackageQuantityWithHttpInfo(basketIdent, packageId, updatePackageQuantityRequest, _options);
        return result.toPromise();
    }
    /**
     * Sets the quantity of the given item in the basket.
     * Updates the quantity of the given package in the basket. The user must be logged in before the quantity can be changed.
     * @param basketIdent The basket identifier.
     * @param packageId The package identifier.
     * @param [updatePackageQuantityRequest]
     */
    updatePackageQuantity(basketIdent, packageId, updatePackageQuantityRequest, _options) {
        const result = this.api.updatePackageQuantity(basketIdent, packageId, updatePackageQuantityRequest, _options);
        return result.toPromise();
    }
    /**
     * Updates a tier.
     * TODO
     * @param token The webstore identifier.
     * @param tierId The tier identifier
     * @param [updateTierRequest]
     */
    updateTierWithHttpInfo(token, tierId, updateTierRequest, _options) {
        const result = this.api.updateTierWithHttpInfo(token, tierId, updateTierRequest, _options);
        return result.toPromise();
    }
    /**
     * Updates a tier.
     * TODO
     * @param token The webstore identifier.
     * @param tierId The tier identifier
     * @param [updateTierRequest]
     */
    updateTier(token, tierId, updateTierRequest, _options) {
        const result = this.api.updateTier(token, tierId, updateTierRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseHeadlessApi = PromiseHeadlessApi;
