// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/model/basketPackage.js
"use strict";
/**
 * Tebex Headless API
 * The headless API is designed for implementing your own store frontend with the data of your store. You are able to call the Headless API directly from a web browser (such as within an SPA), or from a backend server, such as for in-game GUIs.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: tebex-integrations@overwolf.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasketPackage = void 0;
/**
* A package within an existing basket.
*/
class BasketPackage {
    static getAttributeTypeMap() {
        return BasketPackage.attributeTypeMap;
    }
}
exports.BasketPackage = BasketPackage;
BasketPackage.discriminator = undefined;
BasketPackage.attributeTypeMap = [
    {
        "name": "qty",
        "baseName": "qty",
        "type": "number"
    },
    {
        "name": "type",
        "baseName": "type",
        "type": "BasketPackage.TypeEnum"
    },
    {
        "name": "revenueShare",
        "baseName": "revenue_share",
        "type": "Array<RevenueShare>"
    }
];
(function (BasketPackage) {
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["Single"] = 'single'] = "Single";
        TypeEnum[TypeEnum["Subscription"] = 'subscription'] = "Subscription";
    })(TypeEnum = BasketPackage.TypeEnum || (BasketPackage.TypeEnum = {}));
})(BasketPackage || (exports.BasketPackage = BasketPackage = {}));
