// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/model/category.js
"use strict";
/**
 * Tebex Headless API
 * The headless API is designed for implementing your own store frontend with the data of your store. You are able to call the Headless API directly from a web browser (such as within an SPA), or from a backend server, such as for in-game GUIs.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: tebex-integrations@overwolf.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Category = void 0;
class Category {
    static getAttributeTypeMap() {
        return Category.attributeTypeMap;
    }
}
exports.Category = Category;
Category.discriminator = undefined;
Category.attributeTypeMap = [
    {
        "name": "id",
        "baseName": "id",
        "type": "number"
    },
    {
        "name": "name",
        "baseName": "name",
        "type": "string"
    },
    {
        "name": "slug",
        "baseName": "slug",
        "type": "string"
    },
    {
        "name": "parent",
        "baseName": "parent",
        "type": "object"
    },
    {
        "name": "description",
        "baseName": "description",
        "type": "string"
    },
    {
        "name": "packages",
        "baseName": "packages",
        "type": "Array<Package>"
    },
    {
        "name": "order",
        "baseName": "order",
        "type": "number"
    },
    {
        "name": "displayType",
        "baseName": "display_type",
        "type": "Category.DisplayTypeEnum"
    }
];
(function (Category) {
    let DisplayTypeEnum;
    (function (DisplayTypeEnum) {
        DisplayTypeEnum[DisplayTypeEnum["List"] = 'list'] = "List";
        DisplayTypeEnum[DisplayTypeEnum["Grid"] = 'grid'] = "Grid";
    })(DisplayTypeEnum = Category.DisplayTypeEnum || (Category.DisplayTypeEnum = {}));
})(Category || (exports.Category = Category = {}));
