// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/model/models.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoidAuth = exports.OAuth = exports.ApiKeyAuth = exports.HttpBearerAuth = exports.HttpBasicAuth = exports.ObjectSerializer = void 0;
__exportStar(require("./addBasketPackageRequest"), exports);
__exportStar(require("./applyCreatorCodeRequest"), exports);
__exportStar(require("./basket"), exports);
__exportStar(require("./basketAuthResponseInner"), exports);
__exportStar(require("./basketLinks"), exports);
__exportStar(require("./basketPackage"), exports);
__exportStar(require("./basketResponse"), exports);
__exportStar(require("./cMSPage"), exports);
__exportStar(require("./cMSPagesResponse"), exports);
__exportStar(require("./category"), exports);
__exportStar(require("./categoryResponse"), exports);
__exportStar(require("./coupon"), exports);
__exportStar(require("./createBasketRequest"), exports);
__exportStar(require("./errorResponse"), exports);
__exportStar(require("./giftCard"), exports);
__exportStar(require("./package"), exports);
__exportStar(require("./packageCategory"), exports);
__exportStar(require("./packageResponse"), exports);
__exportStar(require("./removeBasketPackageRequest"), exports);
__exportStar(require("./removeGiftCardRequest"), exports);
__exportStar(require("./revenueShare"), exports);
__exportStar(require("./updatePackageQuantityRequest"), exports);
__exportStar(require("./updateTierRequest"), exports);
__exportStar(require("./updateTierResponse"), exports);
__exportStar(require("./webstore"), exports);
__exportStar(require("./webstoreResponse"), exports);
__exportStar(require("./webstoreResponseData"), exports);
const addBasketPackageRequest_1 = require("./addBasketPackageRequest");
const applyCreatorCodeRequest_1 = require("./applyCreatorCodeRequest");
const basket_1 = require("./basket");
const basketAuthResponseInner_1 = require("./basketAuthResponseInner");
const basketLinks_1 = require("./basketLinks");
const basketPackage_1 = require("./basketPackage");
const basketResponse_1 = require("./basketResponse");
const cMSPage_1 = require("./cMSPage");
const cMSPagesResponse_1 = require("./cMSPagesResponse");
const category_1 = require("./category");
const categoryResponse_1 = require("./categoryResponse");
const coupon_1 = require("./coupon");
const createBasketRequest_1 = require("./createBasketRequest");
const errorResponse_1 = require("./errorResponse");
const giftCard_1 = require("./giftCard");
const package_1 = require("./package");
const packageCategory_1 = require("./packageCategory");
const packageResponse_1 = require("./packageResponse");
const removeBasketPackageRequest_1 = require("./removeBasketPackageRequest");
const removeGiftCardRequest_1 = require("./removeGiftCardRequest");
const revenueShare_1 = require("./revenueShare");
const updatePackageQuantityRequest_1 = require("./updatePackageQuantityRequest");
const updateTierRequest_1 = require("./updateTierRequest");
const updateTierResponse_1 = require("./updateTierResponse");
const webstore_1 = require("./webstore");
const webstoreResponse_1 = require("./webstoreResponse");
const webstoreResponseData_1 = require("./webstoreResponseData");
/* tslint:disable:no-unused-variable */
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
let enumsMap = {
    "BasketPackage.TypeEnum": basketPackage_1.BasketPackage.TypeEnum,
    "Category.DisplayTypeEnum": category_1.Category.DisplayTypeEnum,
};
let typeMap = {
    "AddBasketPackageRequest": addBasketPackageRequest_1.AddBasketPackageRequest,
    "ApplyCreatorCodeRequest": applyCreatorCodeRequest_1.ApplyCreatorCodeRequest,
    "Basket": basket_1.Basket,
    "BasketAuthResponseInner": basketAuthResponseInner_1.BasketAuthResponseInner,
    "BasketLinks": basketLinks_1.BasketLinks,
    "BasketPackage": basketPackage_1.BasketPackage,
    "BasketResponse": basketResponse_1.BasketResponse,
    "CMSPage": cMSPage_1.CMSPage,
    "CMSPagesResponse": cMSPagesResponse_1.CMSPagesResponse,
    "Category": category_1.Category,
    "CategoryResponse": categoryResponse_1.CategoryResponse,
    "Coupon": coupon_1.Coupon,
    "CreateBasketRequest": createBasketRequest_1.CreateBasketRequest,
    "ErrorResponse": errorResponse_1.ErrorResponse,
    "GiftCard": giftCard_1.GiftCard,
    "Package": package_1.Package,
    "PackageCategory": packageCategory_1.PackageCategory,
    "PackageResponse": packageResponse_1.PackageResponse,
    "RemoveBasketPackageRequest": removeBasketPackageRequest_1.RemoveBasketPackageRequest,
    "RemoveGiftCardRequest": removeGiftCardRequest_1.RemoveGiftCardRequest,
    "RevenueShare": revenueShare_1.RevenueShare,
    "UpdatePackageQuantityRequest": updatePackageQuantityRequest_1.UpdatePackageQuantityRequest,
    "UpdateTierRequest": updateTierRequest_1.UpdateTierRequest,
    "UpdateTierResponse": updateTierResponse_1.UpdateTierResponse,
    "Webstore": webstore_1.Webstore,
    "WebstoreResponse": webstoreResponse_1.WebstoreResponse,
    "WebstoreResponseData": webstoreResponseData_1.WebstoreResponseData,
};
// Check if a string starts with another string without using es6 features
function startsWith(str, match) {
    return str.substring(0, match.length) === match;
}
// Check if a string ends with another string without using es6 features
function endsWith(str, match) {
    return str.length >= match.length && str.substring(str.length - match.length) === match;
}
const nullableSuffix = " | null";
const optionalSuffix = " | undefined";
const arrayPrefix = "Array<";
const arraySuffix = ">";
const mapPrefix = "{ [key: string]: ";
const mapSuffix = "; }";
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap[expectedType]) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType; // w/e we don't know the type
            }
            // Check the discriminator
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType; // the type does not have a discriminator. use it.
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType; // use the type given in the discriminator
                    }
                    else {
                        return expectedType; // discriminator did not map to a type
                    }
                }
                else {
                    return expectedType; // discriminator was not present (or an empty string)
                }
            }
        }
    }
    static serialize(data, type) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (endsWith(type, nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.serialize(data, subType);
        }
        else if (endsWith(type, optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.serialize(data, subType);
        }
        else if (startsWith(type, arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let index = 0; index < data.length; index++) {
                let datum = data[index];
                transformedData.push(ObjectSerializer.serialize(datum, subType));
            }
            return transformedData;
        }
        else if (startsWith(type, mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.serialize(data[key], subType);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return data.toISOString();
        }
        else {
            if (enumsMap[type]) {
                return data;
            }
            if (!typeMap[type]) { // in case we dont know the type
                return data;
            }
            // Get the actual type of this object
            type = this.findCorrectType(data, type);
            // get the map for the correct type.
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index = 0; index < attributeTypes.length; index++) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type);
            }
            return instance;
        }
    }
    static deserialize(data, type) {
        // polymorphism may change the actual type.
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (endsWith(type, nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.deserialize(data, subType);
        }
        else if (endsWith(type, optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.deserialize(data, subType);
        }
        else if (startsWith(type, arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let index = 0; index < data.length; index++) {
                let datum = data[index];
                transformedData.push(ObjectSerializer.deserialize(datum, subType));
            }
            return transformedData;
        }
        else if (startsWith(type, mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.deserialize(data[key], subType);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap[type]) { // is Enum
                return data;
            }
            if (!typeMap[type]) { // dont know the type
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index = 0; index < attributeTypes.length; index++) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type);
            }
            return instance;
        }
    }
}
exports.ObjectSerializer = ObjectSerializer;
class HttpBasicAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(requestOptions) {
        requestOptions.auth = {
            username: this.username, password: this.password
        };
    }
}
exports.HttpBasicAuth = HttpBasicAuth;
class HttpBearerAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            const accessToken = typeof this.accessToken === 'function'
                ? this.accessToken()
                : this.accessToken;
            requestOptions.headers["Authorization"] = "Bearer " + accessToken;
        }
    }
}
exports.HttpBearerAuth = HttpBearerAuth;
class ApiKeyAuth {
    constructor(location, paramName) {
        this.location = location;
        this.paramName = paramName;
        this.apiKey = '';
    }
    applyToRequest(requestOptions) {
        if (this.location == "query") {
            requestOptions.qs[this.paramName] = this.apiKey;
        }
        else if (this.location == "header" && requestOptions && requestOptions.headers) {
            requestOptions.headers[this.paramName] = this.apiKey;
        }
        else if (this.location == 'cookie' && requestOptions && requestOptions.headers) {
            if (requestOptions.headers['Cookie']) {
                requestOptions.headers['Cookie'] += '; ' + this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
            else {
                requestOptions.headers['Cookie'] = this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
        }
    }
}
exports.ApiKeyAuth = ApiKeyAuth;
class OAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            requestOptions.headers["Authorization"] = "Bearer " + this.accessToken;
        }
    }
}
exports.OAuth = OAuth;
class VoidAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(_) {
        // Do nothing
    }
}
exports.VoidAuth = VoidAuth;
