// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/TebexCheckout/apis/PaymentsApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentsApiResponseProcessor = exports.PaymentsApiRequestFactory = void 0;
// TODO: better import syntax?
const baseapi_1 = require("./baseapi");
const http_1 = require("../http/http");
const ObjectSerializer_1 = require("../models/ObjectSerializer");
const exception_1 = require("./exception");
const util_1 = require("../util");
/**
 * no description
 */
class PaymentsApiRequestFactory extends baseapi_1.BaseAPIRequestFactory {
    /**
     * This will fetch the given payment associated with this transaction id. Single payment transaction IDs begin with `tbx-`
     * Fetch a payment by its transaction ID
     * @param txnId The payment reference to fetch.
     */
    async getPaymentById(txnId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'txnId' is not null or undefined
        if (txnId === null || txnId === undefined) {
            throw new baseapi_1.RequiredError("PaymentsApi", "getPaymentById", "txnId");
        }
        // Path Params
        const localVarPath = '/payments/{txnId}?type=txn_id'
            .replace('{' + 'txnId' + '}', encodeURIComponent(String(txnId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * This will refund the given payment associated with this transaction id.
     * Refund a payment by its transaction ID
     * @param txnId The payment reference to refund.
     */
    async refundPaymentById(txnId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'txnId' is not null or undefined
        if (txnId === null || txnId === undefined) {
            throw new baseapi_1.RequiredError("PaymentsApi", "refundPaymentById", "txnId");
        }
        // Path Params
        const localVarPath = '/payments/{txnId}/refund?type=txn_id'
            .replace('{' + 'txnId' + '}', encodeURIComponent(String(txnId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        let authMethod;
        // Apply auth methods
        authMethod = _config.authMethods["tebex_checkout_auth_basic"];
        if (authMethod?.applySecurityAuthentication) {
            await authMethod?.applySecurityAuthentication(requestContext);
        }
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
}
exports.PaymentsApiRequestFactory = PaymentsApiRequestFactory;
class PaymentsApiResponseProcessor {
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getPaymentById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getPaymentByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Payment", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Transaction not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Payment", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to refundPaymentById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async refundPaymentByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Payment", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The payment cannot be refunded. Payments can only be refunded when a payment\&#39;s &#x60;status&#x60; is Completed.", undefined, response.headers);
        }
        if ((0, util_1.isCodeInRange)("404", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "Payment not found.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Payment", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
}
exports.PaymentsApiResponseProcessor = PaymentsApiResponseProcessor;
