// @tebexio/tebex-sdk-nodejs/dist/Webhook/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookStatus = exports.WebhookType = void 0;
exports.getWebhookType = getWebhookType;
var WebhookType;
(function (WebhookType) {
    WebhookType["PAYMENT_COMPLETED"] = "payment.completed";
    WebhookType["PAYMENT_DECLINED"] = "payment.declined";
    WebhookType["PAYMENT_REFUNDED"] = "payment.refunded";
    WebhookType["PAYMENT_DISPUTE_OPENED"] = "payment.dispute.opened";
    WebhookType["PAYMENT_DISPUTE_WON"] = "payment.dispute.won";
    WebhookType["PAYMENT_DISPUTE_LOST"] = "payment.dispute.lost";
    WebhookType["PAYMENT_DISPUTE_CLOSED"] = "payment.dispute.closed";
    WebhookType["RECURRING_PAYMENT_STARTED"] = "recurring-payment.started";
    WebhookType["RECURRING_PAYMENT_RENEWED"] = "recurring-payment.renewed";
    WebhookType["RECURRING_PAYMENT_STATUS_CHANGED"] = "recurring-payment.status-changed";
    WebhookType["RECURRING_PAYMENT_ENDED"] = "recurring-payment.ended";
    WebhookType["RECURRING_PAYMENT_CANCELLATION_REQUESTED"] = "recurring-payment.cancellation.requested";
    WebhookType["RECURRING_PAYMENT_CANCELLATION_ABORTED"] = "recurring-payment.cancellation.aborted";
    WebhookType["BASKET_ABANDONED"] = "basket.abandoned";
    WebhookType["VALIDATION_WEBHOOK"] = "validation.webhook";
})(WebhookType || (exports.WebhookType = WebhookType = {}));
function getWebhookType(input) {
    return Object.values(WebhookType).find((type) => type === input);
}
// Possible webhook statuses
var WebhookStatus;
(function (WebhookStatus) {
    WebhookStatus[WebhookStatus["COMPLETE"] = 1] = "COMPLETE";
    WebhookStatus[WebhookStatus["REFUND"] = 2] = "REFUND";
    WebhookStatus[WebhookStatus["CHARGEBACK"] = 3] = "CHARGEBACK";
    WebhookStatus[WebhookStatus["DECLINED"] = 18] = "DECLINED";
    WebhookStatus[WebhookStatus["PENDING_CHECKOUT"] = 19] = "PENDING_CHECKOUT";
    WebhookStatus[WebhookStatus["REFUND_PENDING"] = 21] = "REFUND_PENDING";
})(WebhookStatus || (exports.WebhookStatus = WebhookStatus = {}));
