// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("../models/AddPackageRequest"), exports);
__exportStar(require("../models/Address"), exports);
__exportStar(require("../models/Basket"), exports);
__exportStar(require("../models/BasketItem"), exports);
__exportStar(require("../models/BasketLinks"), exports);
__exportStar(require("../models/BasketRow"), exports);
__exportStar(require("../models/BasketRowMeta"), exports);
__exportStar(require("../models/BasketRowMetaLimits"), exports);
__exportStar(require("../models/BasketRowMetaLimitsUser"), exports);
__exportStar(require("../models/CheckoutItem"), exports);
__exportStar(require("../models/CheckoutRequest"), exports);
__exportStar(require("../models/CheckoutRequestBasket"), exports);
__exportStar(require("../models/CreateBasketRequest"), exports);
__exportStar(require("../models/ErrorResponse"), exports);
__exportStar(require("../models/Package"), exports);
__exportStar(require("../models/Payment"), exports);
__exportStar(require("../models/PaymentCustomer"), exports);
__exportStar(require("../models/PaymentFees"), exports);
__exportStar(require("../models/PaymentFeesGateway"), exports);
__exportStar(require("../models/PaymentFeesTax"), exports);
__exportStar(require("../models/PaymentPaymentMethod"), exports);
__exportStar(require("../models/PaymentPrice"), exports);
__exportStar(require("../models/PaymentProductsInner"), exports);
__exportStar(require("../models/PaymentProductsInnerBasePrice"), exports);
__exportStar(require("../models/PaymentProductsInnerPaidPrice"), exports);
__exportStar(require("../models/PaymentStatus"), exports);
__exportStar(require("../models/PaymentSubject"), exports);
__exportStar(require("../models/PaymentSubjectCustomer"), exports);
__exportStar(require("../models/PaymentSubjectFees"), exports);
__exportStar(require("../models/PaymentSubjectPaymentMethod"), exports);
__exportStar(require("../models/PaymentSubjectPrice"), exports);
__exportStar(require("../models/PaymentSubjectProductsInner"), exports);
__exportStar(require("../models/PriceDetails"), exports);
__exportStar(require("../models/RecurringPayment"), exports);
__exportStar(require("../models/RecurringPaymentAmount"), exports);
__exportStar(require("../models/RecurringPaymentLinks"), exports);
__exportStar(require("../models/RecurringPaymentStatus"), exports);
__exportStar(require("../models/RecurringPaymentSubject"), exports);
__exportStar(require("../models/RecurringPaymentSubjectPrice"), exports);
__exportStar(require("../models/RecurringPaymentSubjectStatus"), exports);
__exportStar(require("../models/RevenueShare"), exports);
__exportStar(require("../models/Sale"), exports);
__exportStar(require("../models/TebexWebhook"), exports);
__exportStar(require("../models/UpdateRecurringPaymentRequest"), exports);
__exportStar(require("../models/UpdateSubscriptionRequest"), exports);
__exportStar(require("../models/UpdateSubscriptionRequestItemsInner"), exports);
const AddPackageRequest_1 = require("../models/AddPackageRequest");
const Address_1 = require("../models/Address");
const Basket_1 = require("../models/Basket");
const BasketItem_1 = require("../models/BasketItem");
const BasketLinks_1 = require("../models/BasketLinks");
const BasketRow_1 = require("../models/BasketRow");
const BasketRowMeta_1 = require("../models/BasketRowMeta");
const BasketRowMetaLimits_1 = require("../models/BasketRowMetaLimits");
const BasketRowMetaLimitsUser_1 = require("../models/BasketRowMetaLimitsUser");
const CheckoutItem_1 = require("../models/CheckoutItem");
const CheckoutRequest_1 = require("../models/CheckoutRequest");
const CheckoutRequestBasket_1 = require("../models/CheckoutRequestBasket");
const CreateBasketRequest_1 = require("../models/CreateBasketRequest");
const ErrorResponse_1 = require("../models/ErrorResponse");
const Package_1 = require("../models/Package");
const Payment_1 = require("../models/Payment");
const PaymentCustomer_1 = require("../models/PaymentCustomer");
const PaymentFees_1 = require("../models/PaymentFees");
const PaymentFeesGateway_1 = require("../models/PaymentFeesGateway");
const PaymentFeesTax_1 = require("../models/PaymentFeesTax");
const PaymentPaymentMethod_1 = require("../models/PaymentPaymentMethod");
const PaymentPrice_1 = require("../models/PaymentPrice");
const PaymentProductsInner_1 = require("../models/PaymentProductsInner");
const PaymentProductsInnerBasePrice_1 = require("../models/PaymentProductsInnerBasePrice");
const PaymentProductsInnerPaidPrice_1 = require("../models/PaymentProductsInnerPaidPrice");
const PaymentStatus_1 = require("../models/PaymentStatus");
const PaymentSubject_1 = require("../models/PaymentSubject");
const PaymentSubjectCustomer_1 = require("../models/PaymentSubjectCustomer");
const PaymentSubjectFees_1 = require("../models/PaymentSubjectFees");
const PaymentSubjectPaymentMethod_1 = require("../models/PaymentSubjectPaymentMethod");
const PaymentSubjectPrice_1 = require("../models/PaymentSubjectPrice");
const PaymentSubjectProductsInner_1 = require("../models/PaymentSubjectProductsInner");
const PriceDetails_1 = require("../models/PriceDetails");
const RecurringPayment_1 = require("../models/RecurringPayment");
const RecurringPaymentAmount_1 = require("../models/RecurringPaymentAmount");
const RecurringPaymentLinks_1 = require("../models/RecurringPaymentLinks");
const RecurringPaymentStatus_1 = require("../models/RecurringPaymentStatus");
const RecurringPaymentSubject_1 = require("../models/RecurringPaymentSubject");
const RecurringPaymentSubjectPrice_1 = require("../models/RecurringPaymentSubjectPrice");
const RecurringPaymentSubjectStatus_1 = require("../models/RecurringPaymentSubjectStatus");
const RevenueShare_1 = require("../models/RevenueShare");
const Sale_1 = require("../models/Sale");
const TebexWebhook_1 = require("../models/TebexWebhook");
const UpdateRecurringPaymentRequest_1 = require("../models/UpdateRecurringPaymentRequest");
const UpdateSubscriptionRequest_1 = require("../models/UpdateSubscriptionRequest");
const UpdateSubscriptionRequestItemsInner_1 = require("../models/UpdateSubscriptionRequestItemsInner");
/* tslint:disable:no-unused-variable */
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
let enumsMap = new Set([
    "AddPackageRequestTypeEnum",
    "BasketItemTypeEnum",
    "PackageTypeEnum",
    "PackageExpiryPeriodEnum",
    "SaleDiscountTypeEnum",
    "UpdateRecurringPaymentRequestStatusEnum",
    "UpdateSubscriptionRequestItemsInnerTypeEnum",
]);
let typeMap = {
    "AddPackageRequest": AddPackageRequest_1.AddPackageRequest,
    "Address": Address_1.Address,
    "Basket": Basket_1.Basket,
    "BasketItem": BasketItem_1.BasketItem,
    "BasketLinks": BasketLinks_1.BasketLinks,
    "BasketRow": BasketRow_1.BasketRow,
    "BasketRowMeta": BasketRowMeta_1.BasketRowMeta,
    "BasketRowMetaLimits": BasketRowMetaLimits_1.BasketRowMetaLimits,
    "BasketRowMetaLimitsUser": BasketRowMetaLimitsUser_1.BasketRowMetaLimitsUser,
    "CheckoutItem": CheckoutItem_1.CheckoutItem,
    "CheckoutRequest": CheckoutRequest_1.CheckoutRequest,
    "CheckoutRequestBasket": CheckoutRequestBasket_1.CheckoutRequestBasket,
    "CreateBasketRequest": CreateBasketRequest_1.CreateBasketRequest,
    "ErrorResponse": ErrorResponse_1.ErrorResponse,
    "Package": Package_1.Package,
    "Payment": Payment_1.Payment,
    "PaymentCustomer": PaymentCustomer_1.PaymentCustomer,
    "PaymentFees": PaymentFees_1.PaymentFees,
    "PaymentFeesGateway": PaymentFeesGateway_1.PaymentFeesGateway,
    "PaymentFeesTax": PaymentFeesTax_1.PaymentFeesTax,
    "PaymentPaymentMethod": PaymentPaymentMethod_1.PaymentPaymentMethod,
    "PaymentPrice": PaymentPrice_1.PaymentPrice,
    "PaymentProductsInner": PaymentProductsInner_1.PaymentProductsInner,
    "PaymentProductsInnerBasePrice": PaymentProductsInnerBasePrice_1.PaymentProductsInnerBasePrice,
    "PaymentProductsInnerPaidPrice": PaymentProductsInnerPaidPrice_1.PaymentProductsInnerPaidPrice,
    "PaymentStatus": PaymentStatus_1.PaymentStatus,
    "PaymentSubject": PaymentSubject_1.PaymentSubject,
    "PaymentSubjectCustomer": PaymentSubjectCustomer_1.PaymentSubjectCustomer,
    "PaymentSubjectFees": PaymentSubjectFees_1.PaymentSubjectFees,
    "PaymentSubjectPaymentMethod": PaymentSubjectPaymentMethod_1.PaymentSubjectPaymentMethod,
    "PaymentSubjectPrice": PaymentSubjectPrice_1.PaymentSubjectPrice,
    "PaymentSubjectProductsInner": PaymentSubjectProductsInner_1.PaymentSubjectProductsInner,
    "PriceDetails": PriceDetails_1.PriceDetails,
    "RecurringPayment": RecurringPayment_1.RecurringPayment,
    "RecurringPaymentAmount": RecurringPaymentAmount_1.RecurringPaymentAmount,
    "RecurringPaymentLinks": RecurringPaymentLinks_1.RecurringPaymentLinks,
    "RecurringPaymentStatus": RecurringPaymentStatus_1.RecurringPaymentStatus,
    "RecurringPaymentSubject": RecurringPaymentSubject_1.RecurringPaymentSubject,
    "RecurringPaymentSubjectPrice": RecurringPaymentSubjectPrice_1.RecurringPaymentSubjectPrice,
    "RecurringPaymentSubjectStatus": RecurringPaymentSubjectStatus_1.RecurringPaymentSubjectStatus,
    "RevenueShare": RevenueShare_1.RevenueShare,
    "Sale": Sale_1.Sale,
    "TebexWebhook": TebexWebhook_1.TebexWebhook,
    "UpdateRecurringPaymentRequest": UpdateRecurringPaymentRequest_1.UpdateRecurringPaymentRequest,
    "UpdateSubscriptionRequest": UpdateSubscriptionRequest_1.UpdateSubscriptionRequest,
    "UpdateSubscriptionRequestItemsInner": UpdateSubscriptionRequestItemsInner_1.UpdateSubscriptionRequestItemsInner,
};
/**
 * Every mime-type consists of a type, subtype, and optional parameters.
 * The subtype can be composite, including information about the content format.
 * For example: `application/json-patch+json`, `application/merge-patch+json`.
 *
 * This helper transforms a string mime-type into an internal representation.
 * This simplifies the implementation of predicates that in turn define common rules for parsing or stringifying
 * the payload.
 */
const parseMimeType = (mimeType) => {
    const [type = '', subtype = ''] = mimeType.split('/');
    return {
        type,
        subtype,
        subtypeTokens: subtype.split('+'),
    };
};
// This factory creates a predicate function that checks a string mime-type against defined rules.
const mimeTypePredicateFactory = (predicate) => (mimeType) => predicate(parseMimeType(mimeType));
// Use this factory when you need to define a simple predicate based only on type and, if applicable, subtype.
const mimeTypeSimplePredicateFactory = (type, subtype) => mimeTypePredicateFactory((descriptor) => {
    if (descriptor.type !== type)
        return false;
    if (subtype != null && descriptor.subtype !== subtype)
        return false;
    return true;
});
// Creating a set of named predicates that will help us determine how to handle different mime-types
const isTextLikeMimeType = mimeTypeSimplePredicateFactory('text');
const isJsonMimeType = mimeTypeSimplePredicateFactory('application', 'json');
const isJsonLikeMimeType = mimeTypePredicateFactory((descriptor) => descriptor.type === 'application' && descriptor.subtypeTokens.some((item) => item === 'json'));
const isOctetStreamMimeType = mimeTypeSimplePredicateFactory('application', 'octet-stream');
const isFormUrlencodedMimeType = mimeTypeSimplePredicateFactory('application', 'x-www-form-urlencoded');
// Defining a list of mime-types in the order of prioritization for handling.
const supportedMimeTypePredicatesWithPriority = [
    isJsonMimeType,
    isJsonLikeMimeType,
    isTextLikeMimeType,
    isOctetStreamMimeType,
    isFormUrlencodedMimeType,
];
const nullableSuffix = " | null";
const optionalSuffix = " | undefined";
const arrayPrefix = "Array<";
const arraySuffix = ">";
const mapPrefix = "{ [key: string]: ";
const mapSuffix = "; }";
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType; // w/e we don't know the type
            }
            // Check the discriminator
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType; // the type does not have a discriminator. use it.
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    let mapping = typeMap[expectedType].mapping;
                    if (mapping != undefined && mapping[discriminatorType]) {
                        return mapping[discriminatorType]; // use the type given in the discriminator
                    }
                    else if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType; // discriminator did not map to a type
                    }
                }
                else {
                    return expectedType; // discriminator was not present (or an empty string)
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.endsWith(nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.serialize(data, subType, format);
        }
        else if (type.endsWith(optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.serialize(data, subType, format);
        }
        else if (type.startsWith(arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let date of data) {
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type.startsWith(mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.serialize(data[key], subType, format);
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) { // in case we dont know the type
                return data;
            }
            // Get the actual type of this object
            type = this.findCorrectType(data, type);
            // get the map for the correct type.
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let attributeType of attributeTypes) {
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        // polymorphism may change the actual type.
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.endsWith(nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.deserialize(data, subType, format);
        }
        else if (type.endsWith(optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.deserialize(data, subType, format);
        }
        else if (type.startsWith(arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let date of data) {
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type.startsWith(mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.deserialize(data[key], subType, format);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) { // is Enum
                return data;
            }
            if (!typeMap[type]) { // dont know the type
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let attributeType of attributeTypes) {
                let value = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
                if (value !== undefined) {
                    instance[attributeType.name] = value;
                }
            }
            return instance;
        }
    }
    /**
     * Normalize media type
     *
     * We currently do not handle any media types attributes, i.e. anything
     * after a semicolon. All content is assumed to be UTF-8 compatible.
     */
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return (mediaType.split(";")[0] ?? '').trim().toLowerCase();
    }
    /**
     * From a list of possible media types, choose the one we can handle best.
     *
     * The order of the given media types does not have any impact on the choice
     * made.
     */
    static getPreferredMediaType(mediaTypes) {
        /** According to OAS 3 we should default to json */
        if (mediaTypes.length === 0) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        for (const predicate of supportedMimeTypePredicatesWithPriority) {
            for (const mediaType of normalMediaTypes) {
                if (mediaType != null && predicate(mediaType)) {
                    return mediaType;
                }
            }
        }
        throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
    }
    /**
     * Convert data to a string according the given media type
     */
    static stringify(data, mediaType) {
        if (isTextLikeMimeType(mediaType)) {
            return String(data);
        }
        if (isJsonLikeMimeType(mediaType)) {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    /**
     * Parse data from a string according to the given media type
     */
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (isTextLikeMimeType(mediaType)) {
            return rawData;
        }
        if (isJsonLikeMimeType(mediaType)) {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
