// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/model/models.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoidAuth = exports.OAuth = exports.ApiKeyAuth = exports.HttpBearerAuth = exports.HttpBasicAuth = exports.ObjectSerializer = void 0;
__exportStar(require("./addPackageRequest"), exports);
__exportStar(require("./address"), exports);
__exportStar(require("./basket"), exports);
__exportStar(require("./basketItem"), exports);
__exportStar(require("./basketLinks"), exports);
__exportStar(require("./basketRow"), exports);
__exportStar(require("./basketRowMeta"), exports);
__exportStar(require("./basketRowMetaLimits"), exports);
__exportStar(require("./basketRowMetaLimitsUser"), exports);
__exportStar(require("./checkoutItem"), exports);
__exportStar(require("./checkoutRequest"), exports);
__exportStar(require("./checkoutRequestBasket"), exports);
__exportStar(require("./createBasketRequest"), exports);
__exportStar(require("./errorResponse"), exports);
__exportStar(require("./package"), exports);
__exportStar(require("./payment"), exports);
__exportStar(require("./paymentCustomer"), exports);
__exportStar(require("./paymentFees"), exports);
__exportStar(require("./paymentFeesGateway"), exports);
__exportStar(require("./paymentFeesTax"), exports);
__exportStar(require("./paymentPaymentMethod"), exports);
__exportStar(require("./paymentPrice"), exports);
__exportStar(require("./paymentProductsInner"), exports);
__exportStar(require("./paymentProductsInnerBasePrice"), exports);
__exportStar(require("./paymentProductsInnerPaidPrice"), exports);
__exportStar(require("./paymentStatus"), exports);
__exportStar(require("./paymentSubject"), exports);
__exportStar(require("./paymentSubjectCustomer"), exports);
__exportStar(require("./paymentSubjectFees"), exports);
__exportStar(require("./paymentSubjectPaymentMethod"), exports);
__exportStar(require("./paymentSubjectPrice"), exports);
__exportStar(require("./paymentSubjectProductsInner"), exports);
__exportStar(require("./priceDetails"), exports);
__exportStar(require("./recurringPayment"), exports);
__exportStar(require("./recurringPaymentAmount"), exports);
__exportStar(require("./recurringPaymentLinks"), exports);
__exportStar(require("./recurringPaymentStatus"), exports);
__exportStar(require("./recurringPaymentSubject"), exports);
__exportStar(require("./recurringPaymentSubjectPrice"), exports);
__exportStar(require("./recurringPaymentSubjectStatus"), exports);
__exportStar(require("./revenueShare"), exports);
__exportStar(require("./sale"), exports);
__exportStar(require("./tebexWebhook"), exports);
__exportStar(require("./updateRecurringPaymentRequest"), exports);
__exportStar(require("./updateSubscriptionRequest"), exports);
__exportStar(require("./updateSubscriptionRequestItemsInner"), exports);
const addPackageRequest_1 = require("./addPackageRequest");
const address_1 = require("./address");
const basket_1 = require("./basket");
const basketItem_1 = require("./basketItem");
const basketLinks_1 = require("./basketLinks");
const basketRow_1 = require("./basketRow");
const basketRowMeta_1 = require("./basketRowMeta");
const basketRowMetaLimits_1 = require("./basketRowMetaLimits");
const basketRowMetaLimitsUser_1 = require("./basketRowMetaLimitsUser");
const checkoutItem_1 = require("./checkoutItem");
const checkoutRequest_1 = require("./checkoutRequest");
const checkoutRequestBasket_1 = require("./checkoutRequestBasket");
const createBasketRequest_1 = require("./createBasketRequest");
const errorResponse_1 = require("./errorResponse");
const package_1 = require("./package");
const payment_1 = require("./payment");
const paymentCustomer_1 = require("./paymentCustomer");
const paymentFees_1 = require("./paymentFees");
const paymentFeesGateway_1 = require("./paymentFeesGateway");
const paymentFeesTax_1 = require("./paymentFeesTax");
const paymentPaymentMethod_1 = require("./paymentPaymentMethod");
const paymentPrice_1 = require("./paymentPrice");
const paymentProductsInner_1 = require("./paymentProductsInner");
const paymentProductsInnerBasePrice_1 = require("./paymentProductsInnerBasePrice");
const paymentProductsInnerPaidPrice_1 = require("./paymentProductsInnerPaidPrice");
const paymentStatus_1 = require("./paymentStatus");
const paymentSubject_1 = require("./paymentSubject");
const paymentSubjectCustomer_1 = require("./paymentSubjectCustomer");
const paymentSubjectFees_1 = require("./paymentSubjectFees");
const paymentSubjectPaymentMethod_1 = require("./paymentSubjectPaymentMethod");
const paymentSubjectPrice_1 = require("./paymentSubjectPrice");
const paymentSubjectProductsInner_1 = require("./paymentSubjectProductsInner");
const priceDetails_1 = require("./priceDetails");
const recurringPayment_1 = require("./recurringPayment");
const recurringPaymentAmount_1 = require("./recurringPaymentAmount");
const recurringPaymentLinks_1 = require("./recurringPaymentLinks");
const recurringPaymentStatus_1 = require("./recurringPaymentStatus");
const recurringPaymentSubject_1 = require("./recurringPaymentSubject");
const recurringPaymentSubjectPrice_1 = require("./recurringPaymentSubjectPrice");
const recurringPaymentSubjectStatus_1 = require("./recurringPaymentSubjectStatus");
const revenueShare_1 = require("./revenueShare");
const sale_1 = require("./sale");
const tebexWebhook_1 = require("./tebexWebhook");
const updateRecurringPaymentRequest_1 = require("./updateRecurringPaymentRequest");
const updateSubscriptionRequest_1 = require("./updateSubscriptionRequest");
const updateSubscriptionRequestItemsInner_1 = require("./updateSubscriptionRequestItemsInner");
/* tslint:disable:no-unused-variable */
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
let enumsMap = {
    "AddPackageRequest.TypeEnum": addPackageRequest_1.AddPackageRequest.TypeEnum,
    "BasketItem.TypeEnum": basketItem_1.BasketItem.TypeEnum,
    "Package.TypeEnum": package_1.Package.TypeEnum,
    "Package.ExpiryPeriodEnum": package_1.Package.ExpiryPeriodEnum,
    "Sale.DiscountTypeEnum": sale_1.Sale.DiscountTypeEnum,
    "UpdateRecurringPaymentRequest.StatusEnum": updateRecurringPaymentRequest_1.UpdateRecurringPaymentRequest.StatusEnum,
    "UpdateSubscriptionRequestItemsInner.TypeEnum": updateSubscriptionRequestItemsInner_1.UpdateSubscriptionRequestItemsInner.TypeEnum,
};
let typeMap = {
    "AddPackageRequest": addPackageRequest_1.AddPackageRequest,
    "Address": address_1.Address,
    "Basket": basket_1.Basket,
    "BasketItem": basketItem_1.BasketItem,
    "BasketLinks": basketLinks_1.BasketLinks,
    "BasketRow": basketRow_1.BasketRow,
    "BasketRowMeta": basketRowMeta_1.BasketRowMeta,
    "BasketRowMetaLimits": basketRowMetaLimits_1.BasketRowMetaLimits,
    "BasketRowMetaLimitsUser": basketRowMetaLimitsUser_1.BasketRowMetaLimitsUser,
    "CheckoutItem": checkoutItem_1.CheckoutItem,
    "CheckoutRequest": checkoutRequest_1.CheckoutRequest,
    "CheckoutRequestBasket": checkoutRequestBasket_1.CheckoutRequestBasket,
    "CreateBasketRequest": createBasketRequest_1.CreateBasketRequest,
    "ErrorResponse": errorResponse_1.ErrorResponse,
    "Package": package_1.Package,
    "Payment": payment_1.Payment,
    "PaymentCustomer": paymentCustomer_1.PaymentCustomer,
    "PaymentFees": paymentFees_1.PaymentFees,
    "PaymentFeesGateway": paymentFeesGateway_1.PaymentFeesGateway,
    "PaymentFeesTax": paymentFeesTax_1.PaymentFeesTax,
    "PaymentPaymentMethod": paymentPaymentMethod_1.PaymentPaymentMethod,
    "PaymentPrice": paymentPrice_1.PaymentPrice,
    "PaymentProductsInner": paymentProductsInner_1.PaymentProductsInner,
    "PaymentProductsInnerBasePrice": paymentProductsInnerBasePrice_1.PaymentProductsInnerBasePrice,
    "PaymentProductsInnerPaidPrice": paymentProductsInnerPaidPrice_1.PaymentProductsInnerPaidPrice,
    "PaymentStatus": paymentStatus_1.PaymentStatus,
    "PaymentSubject": paymentSubject_1.PaymentSubject,
    "PaymentSubjectCustomer": paymentSubjectCustomer_1.PaymentSubjectCustomer,
    "PaymentSubjectFees": paymentSubjectFees_1.PaymentSubjectFees,
    "PaymentSubjectPaymentMethod": paymentSubjectPaymentMethod_1.PaymentSubjectPaymentMethod,
    "PaymentSubjectPrice": paymentSubjectPrice_1.PaymentSubjectPrice,
    "PaymentSubjectProductsInner": paymentSubjectProductsInner_1.PaymentSubjectProductsInner,
    "PriceDetails": priceDetails_1.PriceDetails,
    "RecurringPayment": recurringPayment_1.RecurringPayment,
    "RecurringPaymentAmount": recurringPaymentAmount_1.RecurringPaymentAmount,
    "RecurringPaymentLinks": recurringPaymentLinks_1.RecurringPaymentLinks,
    "RecurringPaymentStatus": recurringPaymentStatus_1.RecurringPaymentStatus,
    "RecurringPaymentSubject": recurringPaymentSubject_1.RecurringPaymentSubject,
    "RecurringPaymentSubjectPrice": recurringPaymentSubjectPrice_1.RecurringPaymentSubjectPrice,
    "RecurringPaymentSubjectStatus": recurringPaymentSubjectStatus_1.RecurringPaymentSubjectStatus,
    "RevenueShare": revenueShare_1.RevenueShare,
    "Sale": sale_1.Sale,
    "TebexWebhook": tebexWebhook_1.TebexWebhook,
    "UpdateRecurringPaymentRequest": updateRecurringPaymentRequest_1.UpdateRecurringPaymentRequest,
    "UpdateSubscriptionRequest": updateSubscriptionRequest_1.UpdateSubscriptionRequest,
    "UpdateSubscriptionRequestItemsInner": updateSubscriptionRequestItemsInner_1.UpdateSubscriptionRequestItemsInner,
};
// Check if a string starts with another string without using es6 features
function startsWith(str, match) {
    return str.substring(0, match.length) === match;
}
// Check if a string ends with another string without using es6 features
function endsWith(str, match) {
    return str.length >= match.length && str.substring(str.length - match.length) === match;
}
const nullableSuffix = " | null";
const optionalSuffix = " | undefined";
const arrayPrefix = "Array<";
const arraySuffix = ">";
const mapPrefix = "{ [key: string]: ";
const mapSuffix = "; }";
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap[expectedType]) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType; // w/e we don't know the type
            }
            // Check the discriminator
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType; // the type does not have a discriminator. use it.
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType; // use the type given in the discriminator
                    }
                    else {
                        return expectedType; // discriminator did not map to a type
                    }
                }
                else {
                    return expectedType; // discriminator was not present (or an empty string)
                }
            }
        }
    }
    static serialize(data, type) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (endsWith(type, nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.serialize(data, subType);
        }
        else if (endsWith(type, optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.serialize(data, subType);
        }
        else if (startsWith(type, arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let index = 0; index < data.length; index++) {
                let datum = data[index];
                transformedData.push(ObjectSerializer.serialize(datum, subType));
            }
            return transformedData;
        }
        else if (startsWith(type, mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.serialize(data[key], subType);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return data.toISOString();
        }
        else {
            if (enumsMap[type]) {
                return data;
            }
            if (!typeMap[type]) { // in case we dont know the type
                return data;
            }
            // Get the actual type of this object
            type = this.findCorrectType(data, type);
            // get the map for the correct type.
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index = 0; index < attributeTypes.length; index++) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type);
            }
            return instance;
        }
    }
    static deserialize(data, type) {
        // polymorphism may change the actual type.
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (endsWith(type, nullableSuffix)) {
            let subType = type.slice(0, -nullableSuffix.length); // Type | null => Type
            return ObjectSerializer.deserialize(data, subType);
        }
        else if (endsWith(type, optionalSuffix)) {
            let subType = type.slice(0, -optionalSuffix.length); // Type | undefined => Type
            return ObjectSerializer.deserialize(data, subType);
        }
        else if (startsWith(type, arrayPrefix)) {
            let subType = type.slice(arrayPrefix.length, -arraySuffix.length); // Array<Type> => Type
            let transformedData = [];
            for (let index = 0; index < data.length; index++) {
                let datum = data[index];
                transformedData.push(ObjectSerializer.deserialize(datum, subType));
            }
            return transformedData;
        }
        else if (startsWith(type, mapPrefix)) {
            let subType = type.slice(mapPrefix.length, -mapSuffix.length); // { [key: string]: Type; } => Type
            let transformedData = {};
            for (let key in data) {
                transformedData[key] = ObjectSerializer.deserialize(data[key], subType);
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap[type]) { // is Enum
                return data;
            }
            if (!typeMap[type]) { // dont know the type
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index = 0; index < attributeTypes.length; index++) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type);
            }
            return instance;
        }
    }
}
exports.ObjectSerializer = ObjectSerializer;
class HttpBasicAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(requestOptions) {
        requestOptions.auth = {
            username: this.username, password: this.password
        };
    }
}
exports.HttpBasicAuth = HttpBasicAuth;
class HttpBearerAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            const accessToken = typeof this.accessToken === 'function'
                ? this.accessToken()
                : this.accessToken;
            requestOptions.headers["Authorization"] = "Bearer " + accessToken;
        }
    }
}
exports.HttpBearerAuth = HttpBearerAuth;
class ApiKeyAuth {
    constructor(location, paramName) {
        this.location = location;
        this.paramName = paramName;
        this.apiKey = '';
    }
    applyToRequest(requestOptions) {
        if (this.location == "query") {
            requestOptions.qs[this.paramName] = this.apiKey;
        }
        else if (this.location == "header" && requestOptions && requestOptions.headers) {
            requestOptions.headers[this.paramName] = this.apiKey;
        }
        else if (this.location == 'cookie' && requestOptions && requestOptions.headers) {
            if (requestOptions.headers['Cookie']) {
                requestOptions.headers['Cookie'] += '; ' + this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
            else {
                requestOptions.headers['Cookie'] = this.paramName + '=' + encodeURIComponent(this.apiKey);
            }
        }
    }
}
exports.ApiKeyAuth = ApiKeyAuth;
class OAuth {
    constructor() {
        this.accessToken = '';
    }
    applyToRequest(requestOptions) {
        if (requestOptions && requestOptions.headers) {
            requestOptions.headers["Authorization"] = "Bearer " + this.accessToken;
        }
    }
}
exports.OAuth = OAuth;
class VoidAuth {
    constructor() {
        this.username = '';
        this.password = '';
    }
    applyToRequest(_) {
        // Do nothing
    }
}
exports.VoidAuth = VoidAuth;
