// @tebexio/tebex-sdk-nodejs/dist/Checkout/lib/model/updateRecurringPaymentRequest.js
"use strict";
/**
 * Tebex Checkout API
 * The Checkout APIs are designed to allow our creators to use the Tebex Checkout flow and payment acceptance capabilities without the need to set up a Tebex-powered webstore. Using these APIs allows you to create baskets with custom products (as opposed to pre-created products on our webstore platform), and send customers directly to the checkout flow to proceed with payment options.  You must receive prior authorisation before the Checkout API is enabled on your account. Please contact customer support or your account manager to discover if you qualify to use the Checkout API before beginning integration.
 *
 * The version of the OpenAPI document: 1.1.2
 * Contact: tebex-integrations@overwolf.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateRecurringPaymentRequest = void 0;
class UpdateRecurringPaymentRequest {
    static getAttributeTypeMap() {
        return UpdateRecurringPaymentRequest.attributeTypeMap;
    }
}
exports.UpdateRecurringPaymentRequest = UpdateRecurringPaymentRequest;
UpdateRecurringPaymentRequest.discriminator = undefined;
UpdateRecurringPaymentRequest.attributeTypeMap = [
    {
        "name": "status",
        "baseName": "status",
        "type": "UpdateRecurringPaymentRequest.StatusEnum"
    },
    {
        "name": "pausedUntil",
        "baseName": "paused_until",
        "type": "string"
    }
];
(function (UpdateRecurringPaymentRequest) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["Paused"] = 'Paused'] = "Paused";
        StatusEnum[StatusEnum["Active"] = 'Active'] = "Active";
    })(StatusEnum = UpdateRecurringPaymentRequest.StatusEnum || (UpdateRecurringPaymentRequest.StatusEnum = {}));
})(UpdateRecurringPaymentRequest || (exports.UpdateRecurringPaymentRequest = UpdateRecurringPaymentRequest = {}));
