// @tebexio/tebex-sdk-nodejs/dist/Headless/lib/TebexHeadless/apis/HeadlessApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeadlessApiResponseProcessor = exports.HeadlessApiRequestFactory = void 0;
// TODO: better import syntax?
const baseapi_1 = require("./baseapi");
const http_1 = require("../http/http");
const ObjectSerializer_1 = require("../models/ObjectSerializer");
const exception_1 = require("./exception");
const util_1 = require("../util");
/**
 * no description
 */
class HeadlessApiRequestFactory extends baseapi_1.BaseAPIRequestFactory {
    /**
     * Add a package with the given ID to the basket.
     * Add a package to a basket
     * @param basketIdent The basket identifier.
     * @param addBasketPackageRequest
     */
    async addBasketPackage(basketIdent, addBasketPackageRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "addBasketPackage", "basketIdent");
        }
        // Path Params
        const localVarPath = '/baskets/{basketIdent}/packages'
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(addBasketPackageRequest, "AddBasketPackageRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Applies a creator code to a basket.
     * Apply a coupon to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param coupon Provide a &#x60;coupon_code&#x60; to apply to the basket.
     */
    async applyCoupon(token, basketIdent, coupon, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyCoupon", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyCoupon", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/coupons'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(coupon, "Coupon", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Applies a creator code to a basket.
     * Apply a creator code to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param applyCreatorCodeRequest Provide a &#x60;creator_code&#x60; to apply to the basket.
     */
    async applyCreatorCode(token, basketIdent, applyCreatorCodeRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyCreatorCode", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyCreatorCode", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/creator-codes'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(applyCreatorCodeRequest, "ApplyCreatorCodeRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Applies a creator code to a basket.
     * Apply a gift card to a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param giftCard Provide a &#x60;card_number&#x60; to apply to the basket.
     */
    async applyGiftCard(token, basketIdent, giftCard, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyGiftCard", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "applyGiftCard", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/giftcards'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(giftCard, "GiftCard", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Creates a new basket for use during checkout.
     * Create a new basket
     * @param token The webstore identifier.
     * @param createBasketRequest
     */
    async createBasket(token, createBasketRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "createBasket", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(createBasketRequest, "CreateBasketRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all categories from a webstore. This does not include package information. To include package information, add `?includePackages=1` to the URL.
     * Gets all categories available in the webstore.
     * @param token The webstore identifier.
     */
    async getAllCategories(token, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllCategories", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/categories'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all categories from a webstore including packages.
     * Gets a store\'s categories including all package information with them.
     * @param token The webstore identifier.
     */
    async getAllCategoriesIncludingPackages(token, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllCategoriesIncludingPackages", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/categories?includePackages=1'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all packages from a webstore.
     * Fetch all packages from a webstore
     * @param token The webstore identifier.
     */
    async getAllPackages(token, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackages", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/packages'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    async getAllPackagesWithAuthedIP(token, ipAddress, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithAuthedIP", "token");
        }
        // verify required parameter 'ipAddress' is not null or undefined
        if (ipAddress === null || ipAddress === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithAuthedIP", "ipAddress");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/packages?ipAddress={ipAddress}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'ipAddress' + '}', encodeURIComponent(String(ipAddress)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param ipAddress An IP address can be provided with authenticated requests.
     */
    async getAllPackagesWithAuthedIPAndBasket(token, basketIdent, ipAddress, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithAuthedIPAndBasket", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithAuthedIPAndBasket", "basketIdent");
        }
        // verify required parameter 'ipAddress' is not null or undefined
        if (ipAddress === null || ipAddress === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithAuthedIPAndBasket", "ipAddress");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/packages?ipAddress={ipAddress}&basketIdent={basketIdent}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)))
            .replace('{' + 'ipAddress' + '}', encodeURIComponent(String(ipAddress)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets all packages from a webstore.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    async getAllPackagesWithBasket(token, basketIdent, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithBasket", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getAllPackagesWithBasket", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/packages?basketIdent={basketIdent}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Fetches a basket\'s auth URL.
     * Get authentication links for a basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param returnUrl The URL you would like to redirect the user to after successful basket authentication.
     */
    async getBasketAuthUrl(token, basketIdent, returnUrl, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getBasketAuthUrl", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getBasketAuthUrl", "basketIdent");
        }
        // verify required parameter 'returnUrl' is not null or undefined
        if (returnUrl === null || returnUrl === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getBasketAuthUrl", "returnUrl");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/auth?returnUrl={returnUrl}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)))
            .replace('{' + 'returnUrl' + '}', encodeURIComponent(String(returnUrl)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets a basket associated with the provided identifier.
     * Fetch a basket from a webstore by its identifier
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    async getBasketById(token, basketIdent, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getBasketById", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getBasketById", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets a list of custom pages associated with the webstore. These contain a `content` variable with the HTML content of the page.
     * Fetch the custom pages associated with the store.
     * @param token The webstore identifier.
     */
    async getCMSPages(token, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getCMSPages", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/pages'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    async getCategoryById(token, categoryId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getCategoryById", "token");
        }
        // verify required parameter 'categoryId' is not null or undefined
        if (categoryId === null || categoryId === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getCategoryById", "categoryId");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/categories/{categoryId}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'categoryId' + '}', encodeURIComponent(String(categoryId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets information about a category and returns the packages in that category.
     * Gets information about a specific category, including all the packages in the category
     * @param token The webstore identifier.
     * @param categoryId The ID of the category to fetch.
     */
    async getCategoryIncludingPackages(token, categoryId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getCategoryIncludingPackages", "token");
        }
        // verify required parameter 'categoryId' is not null or undefined
        if (categoryId === null || categoryId === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getCategoryIncludingPackages", "categoryId");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/categories/{categoryId}?includePackages=1'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'categoryId' + '}', encodeURIComponent(String(categoryId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets a package from a webstore by ID.
     * Fetch a package from a webstore by its identifier
     * @param token The webstore identifier.
     * @param packageId The package\&#39;s ID.
     */
    async getPackageById(token, packageId, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getPackageById", "token");
        }
        // verify required parameter 'packageId' is not null or undefined
        if (packageId === null || packageId === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getPackageById", "packageId");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/packages/{packageId}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'packageId' + '}', encodeURIComponent(String(packageId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Gets the webstore associated with the provided identifier.
     * Fetch a webstore by its identifier
     * @param token The webstore identifier.
     */
    async getWebstoreById(token, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "getWebstoreById", "token");
        }
        // Path Params
        const localVarPath = '/accounts/{token}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.GET);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Remove the given package ID from the basket.
     * Remove a package from a basket
     * @param basketIdent The basket identifier.
     * @param removeBasketPackageRequest
     */
    async removeBasketPackage(basketIdent, removeBasketPackageRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeBasketPackage", "basketIdent");
        }
        // Path Params
        const localVarPath = '/baskets/{basketIdent}/packages/remove'
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(removeBasketPackageRequest, "RemoveBasketPackageRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Removes a coupon code from a basket
     * Remove a coupon from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    async removeCoupon(token, basketIdent, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeCoupon", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeCoupon", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/coupons/remove'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Applies a creator code to a basket.
     * Remove a creator code from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     */
    async removeCreatorCode(token, basketIdent, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeCreatorCode", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeCreatorCode", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/creator-codes/remove'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Removes a gift card from the basket.
     * Remove a gift card from the basket.
     * @param token The webstore identifier.
     * @param basketIdent The basket identifier.
     * @param removeGiftCardRequest Provide the &#x60;card_number&#x60; to remove from the basket.
     */
    async removeGiftCard(token, basketIdent, removeGiftCardRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeGiftCard", "token");
        }
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "removeGiftCard", "basketIdent");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/baskets/{basketIdent}/giftcards/remove'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.POST);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(removeGiftCardRequest, "RemoveGiftCardRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Sets the quantity of the given item in the basket.
     * Updates the quantity of the given package in the basket. The user must be logged in before the quantity can be changed.
     * @param basketIdent The basket identifier.
     * @param packageId The package identifier.
     * @param updatePackageQuantityRequest
     */
    async updatePackageQuantity(basketIdent, packageId, updatePackageQuantityRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'basketIdent' is not null or undefined
        if (basketIdent === null || basketIdent === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "updatePackageQuantity", "basketIdent");
        }
        // verify required parameter 'packageId' is not null or undefined
        if (packageId === null || packageId === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "updatePackageQuantity", "packageId");
        }
        // Path Params
        const localVarPath = '/baskets/{basketIdent}/packages/{packageId}'
            .replace('{' + 'basketIdent' + '}', encodeURIComponent(String(basketIdent)))
            .replace('{' + 'packageId' + '}', encodeURIComponent(String(packageId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.PUT);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(updatePackageQuantityRequest, "UpdatePackageQuantityRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
    /**
     * Updates a tier.
     * TODO
     * @param token The webstore identifier.
     * @param tierId The tier identifier
     * @param updateTierRequest
     */
    async updateTier(token, tierId, updateTierRequest, _options) {
        let _config = _options || this.configuration;
        // verify required parameter 'token' is not null or undefined
        if (token === null || token === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "updateTier", "token");
        }
        // verify required parameter 'tierId' is not null or undefined
        if (tierId === null || tierId === undefined) {
            throw new baseapi_1.RequiredError("HeadlessApi", "updateTier", "tierId");
        }
        // Path Params
        const localVarPath = '/accounts/{token}/tiers/{tierId}'
            .replace('{' + 'token' + '}', encodeURIComponent(String(token)))
            .replace('{' + 'tierId' + '}', encodeURIComponent(String(tierId)));
        // Make Request Context
        const requestContext = _config.baseServer.makeRequestContext(localVarPath, http_1.HttpMethod.PATCH);
        requestContext.setHeaderParam("Accept", "application/json, */*;q=0.8");
        // Body Params
        const contentType = ObjectSerializer_1.ObjectSerializer.getPreferredMediaType([
            "application/json"
        ]);
        requestContext.setHeaderParam("Content-Type", contentType);
        const serializedBody = ObjectSerializer_1.ObjectSerializer.stringify(ObjectSerializer_1.ObjectSerializer.serialize(updateTierRequest, "UpdateTierRequest", ""), contentType);
        requestContext.setBody(serializedBody);
        const defaultAuth = _options?.authMethods?.default || this.configuration?.authMethods?.default;
        if (defaultAuth?.applySecurityAuthentication) {
            await defaultAuth?.applySecurityAuthentication(requestContext);
        }
        return requestContext;
    }
}
exports.HeadlessApiRequestFactory = HeadlessApiRequestFactory;
class HeadlessApiResponseProcessor {
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to addBasketPackage
     * @throws ApiException if the response code was not in [200, 299]
     */
    async addBasketPackageWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to applyCoupon
     * @throws ApiException if the response code was not in [200, 299]
     */
    async applyCouponWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to applyCreatorCode
     * @throws ApiException if the response code was not in [200, 299]
     */
    async applyCreatorCodeWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to applyGiftCard
     * @throws ApiException if the response code was not in [200, 299]
     */
    async applyGiftCardWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to createBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async createBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided webstore ID is invalid.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllCategories
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllCategoriesWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllCategoriesIncludingPackages
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllCategoriesIncludingPackagesWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllPackages
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllPackagesWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllPackagesWithAuthedIP
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllPackagesWithAuthedIPWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllPackagesWithAuthedIPAndBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllPackagesWithAuthedIPAndBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getAllPackagesWithBasket
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getAllPackagesWithBasketWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getBasketAuthUrl
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getBasketAuthUrlWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Array<BasketAuthResponseInner>", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided webstore ID or basket ID is invalid.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Array<BasketAuthResponseInner>", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getBasketById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getBasketByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "BasketResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getCMSPages
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getCMSPagesWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CMSPagesResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CMSPagesResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getCategoryById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getCategoryByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getCategoryIncludingPackages
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getCategoryIncludingPackagesWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CategoryResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getPackageById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getPackageByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "PackageResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to getWebstoreById
     * @throws ApiException if the response code was not in [200, 299]
     */
    async getWebstoreByIdWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "WebstoreResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "WebstoreResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to removeBasketPackage
     * @throws ApiException if the response code was not in [200, 299]
     */
    async removeBasketPackageWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "Basket", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to removeCoupon
     * @throws ApiException if the response code was not in [200, 299]
     */
    async removeCouponWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, undefined);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "void", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to removeCreatorCode
     * @throws ApiException if the response code was not in [200, 299]
     */
    async removeCreatorCodeWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, undefined);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "void", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to removeGiftCard
     * @throws ApiException if the response code was not in [200, 299]
     */
    async removeGiftCardWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, undefined);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "void", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to updatePackageQuantity
     * @throws ApiException if the response code was not in [200, 299]
     */
    async updatePackageQuantityWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, undefined);
        }
        if ((0, util_1.isCodeInRange)("422", response.httpStatusCode)) {
            throw new exception_1.ApiException(response.httpStatusCode, "The provided request is invalid. The error response will include detail as to which parameter failed validation.", undefined, response.headers);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "void", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
    /**
     * Unwraps the actual response sent by the server from the response context and deserializes the response content
     * to the expected objects
     *
     * @params response Response returned by the server for a request to updateTier
     * @throws ApiException if the response code was not in [200, 299]
     */
    async updateTierWithHttpInfo(response) {
        const contentType = ObjectSerializer_1.ObjectSerializer.normalizeMediaType(response.headers["content-type"]);
        if ((0, util_1.isCodeInRange)("200", response.httpStatusCode)) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CMSPagesResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        // Work around for missing responses in specification, e.g. for petstore.yaml
        if (response.httpStatusCode >= 200 && response.httpStatusCode <= 299) {
            const body = ObjectSerializer_1.ObjectSerializer.deserialize(ObjectSerializer_1.ObjectSerializer.parse(await response.body.text(), contentType), "CMSPagesResponse", "");
            return new http_1.HttpInfo(response.httpStatusCode, response.headers, response.body, body);
        }
        throw new exception_1.ApiException(response.httpStatusCode, "Unknown API Status Code!", await response.getBodyAsAny(), response.headers);
    }
}
exports.HeadlessApiResponseProcessor = HeadlessApiResponseProcessor;
