// @tebexio/tebex-sdk-nodejs/dist/Webhooks/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WEBHOOK_STATUSES = exports.WEBHOOK_TYPES = exports.VALIDATION_WEBHOOK = exports.BASKET_ABANDONED = exports.RECURRING_PAYMENT_CANCELLATION_ABORTED = exports.RECURRING_PAYMENT_CANCELLATION_REQUESTED = exports.RECURRING_PAYMENT_ENDED = exports.RECURRING_PAYMENT_STATUS_CHANGED = exports.RECURRING_PAYMENT_RENEWED = exports.RECURRING_PAYMENT_STARTED = exports.PAYMENT_DISPUTE_CLOSED = exports.PAYMENT_DISPUTE_LOST = exports.PAYMENT_DISPUTE_WON = exports.PAYMENT_DISPUTE_OPENED = exports.PAYMENT_REFUNDED = exports.PAYMENT_DECLINED = exports.PAYMENT_COMPLETED = void 0;
// Constant values for webhook types
exports.PAYMENT_COMPLETED = "payment.completed";
exports.PAYMENT_DECLINED = "payment.declined";
exports.PAYMENT_REFUNDED = "payment.refunded";
exports.PAYMENT_DISPUTE_OPENED = "payment.dispute.opened";
exports.PAYMENT_DISPUTE_WON = "payment.dispute.won";
exports.PAYMENT_DISPUTE_LOST = "payment.dispute.lost";
exports.PAYMENT_DISPUTE_CLOSED = "payment.dispute.closed";
exports.RECURRING_PAYMENT_STARTED = "recurring-payment.started";
exports.RECURRING_PAYMENT_RENEWED = "recurring-payment.renewed";
exports.RECURRING_PAYMENT_STATUS_CHANGED = "recurring-payment.status-changed";
exports.RECURRING_PAYMENT_ENDED = "recurring-payment.ended";
exports.RECURRING_PAYMENT_CANCELLATION_REQUESTED = "recurring-payment.cancellation.requested";
exports.RECURRING_PAYMENT_CANCELLATION_ABORTED = "recurring-payment.cancellation.aborted";
exports.BASKET_ABANDONED = "basket.abandoned";
exports.VALIDATION_WEBHOOK = "validation.webhook";
// Map webhook types to their appropriate class
exports.WEBHOOK_TYPES = {
    [exports.PAYMENT_COMPLETED]: "PaymentCompletedWebhook",
    [exports.PAYMENT_DECLINED]: "PaymentDeclinedWebhook",
    [exports.PAYMENT_REFUNDED]: "PaymentRefundedWebhook",
    [exports.PAYMENT_DISPUTE_OPENED]: "PaymentDisputeOpenedWebhook",
    [exports.PAYMENT_DISPUTE_WON]: "PaymentDisputeWonWebhook",
    [exports.PAYMENT_DISPUTE_LOST]: "PaymentDisputeLostWebhook",
    [exports.PAYMENT_DISPUTE_CLOSED]: "PaymentDisputeClosedWebhook",
    [exports.RECURRING_PAYMENT_STARTED]: "RecurringPaymentStartedWebhook",
    [exports.RECURRING_PAYMENT_RENEWED]: "RecurringPaymentRenewed",
    [exports.RECURRING_PAYMENT_STATUS_CHANGED]: "RecurringPaymentStatusChangedWebhook",
    [exports.RECURRING_PAYMENT_ENDED]: "RecurringPaymentEndedWebhook",
    [exports.RECURRING_PAYMENT_CANCELLATION_REQUESTED]: "RecurringPaymentCancellationRequestedWebhook",
    [exports.RECURRING_PAYMENT_CANCELLATION_ABORTED]: "RecurringPaymentCancellationAbortedWebhook",
    [exports.BASKET_ABANDONED]: "BasketAbandonedWebhook",
    [exports.VALIDATION_WEBHOOK]: "ValidationWebhook",
};
// Possible webhook statuses
exports.WEBHOOK_STATUSES = {
    1: "Complete",
    2: "Refund",
    3: "Chargeback",
    18: "Declined",
    19: "Pending Checkout",
    21: "Refund Pending",
};
