// @tebexio/tebex-sdk-nodejs/dist/Webhooks/webhook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = void 0;
const tebex_1 = require("../tebex");
const types_1 = require("./types");
/**
 * Base class inherited by all webhooks. Create a webhook instance using `Webhook.fromJson()` with the received JSON data from Tebex.
 */
class Webhook {
    constructor(rawJson) {
        this._rawJson = rawJson;
        const decodedJson = JSON.parse(rawJson);
        if (!decodedJson) {
            throw new Error(`Invalid or malformed webhook JSON: ${rawJson}`);
        }
        this._encodedJson = JSON.stringify(decodedJson);
        this._id = decodedJson["id"];
        this._type = decodedJson["type"];
        this._date = decodedJson["date"];
        const encodedSubject = JSON.stringify(decodedJson["subject"]);
        if (!encodedSubject) {
            throw new Error(`Invalid or malformed webhook subject: ${JSON.stringify(decodedJson["subject"])}`);
        }
        const decodedSubjectObject = JSON.parse(encodedSubject);
        // Handle empty subject objects which can be parsed as arrays
        if (Array.isArray(decodedSubjectObject) && decodedSubjectObject.length === 0) {
            this._subject = {};
        }
        else {
            this._subject = decodedSubjectObject;
        }
        if (this.isTypeOfPayment() || this.isTypeOfDispute()) {
            this._subject = decodedJson["subject"];
        }
        else if (this.isTypeOfRecurringPayment()) {
            this._subject = decodedJson["subject"];
        }
        else if (this.isType(types_1.VALIDATION_WEBHOOK)) {
            this._subject = {};
        }
    }
    /**
     * Parses and validates the webhook JSON payload.
     *
     * @param webhookJsonStr - The JSON-formatted webhook string payload. If null, reads from process standard input.
     * @param headers - The incoming headers from a webhook request.
     * @returns An instance of the correct Webhook class.
     * @throws ValueError When JSON is invalid, malformed, or missing required data.
     * @throws ApiException For invalid IP or signature validation errors.
     */
    static parse(webhookJsonStr, headers) {
        const json = webhookJsonStr;
        const signature = headers["HTTP_X_SIGNATURE"];
        if (!signature) {
            throw new Error("X_SIGNATURE header is missing from the request");
        }
        const decodedJson = JSON.parse(json);
        if (!decodedJson) {
            throw new Error(`Invalid or malformed webhook JSON: ${json}`);
        }
        const webhookType = decodedJson["type"] ?? null;
        if (!webhookType) {
            throw new Error(`Webhook type is missing from the payload: ${json}`);
        }
        if (!decodedJson.hasOwnProperty("subject")) {
            throw new Error(`Webhook is missing subject from the payload: ${json}`);
        }
        const webhookSubject = decodedJson["subject"];
        if (webhookSubject === null) {
            throw new Error(`Webhook subject is null in payload: ${json}`);
        }
        if (!(webhookType in types_1.WEBHOOK_TYPES)) {
            throw new Error(`Unrecognized webhook type: ${webhookType}`);
        }
        const webhookClass = types_1.WEBHOOK_TYPES[webhookType];
        //FIXME polymorphic return type
        const webhook = new Webhook(json);
        const webhookIp = headers["HTTP_X_FORWARDED_FOR"] ?? headers["REMOTE_ADDR"];
        if (!tebex_1.Tebex.webhooks.validateWebhookIp(webhook, webhookIp)) {
            throw new Error("Invalid webhook origin IP. Ensure client IPs are forwarded appropriately if using a proxy.");
        }
        if (!tebex_1.Tebex.webhooks.validateWebhookSignature(webhook, signature)) {
            throw new Error("Invalid webhook signature");
        }
        return webhook;
    }
    getId() {
        return this._id;
    }
    getDate() {
        return this._date;
    }
    getSubject() {
        return this._subject;
    }
    getType() {
        return this._type;
    }
    isTypeOfPayment() {
        return this._type.includes("payment") && !this._type.includes("recurring");
    }
    isTypeOfDispute() {
        return this._type.includes("dispute");
    }
    isTypeOfRecurringPayment() {
        return this._type.includes("recurring-payment");
    }
    isType(type) {
        if (!(type in types_1.WEBHOOK_TYPES)) {
            throw new Error(`Invalid webhook type: ` + type);
        }
        return this._type === type;
    }
    validateIp(incomingRequestIp) {
        return incomingRequestIp === "18.209.80.3" || incomingRequestIp === "54.87.231.232";
    }
    validateSignature(expectedSignature, webhookSecret) {
        const calculatedSignature = require("crypto")
            .createHmac("sha256", webhookSecret)
            .update(require("crypto").createHash("sha256").update(this._encodedJson).digest("hex"))
            .digest("hex");
        return calculatedSignature === expectedSignature;
    }
    isStatusComplete() {
        const statusId = this._subject?.getStatus?.()["id"];
        return types_1.WEBHOOK_STATUSES[statusId] === "Complete";
    }
}
exports.Webhook = Webhook;
